// @myriaddreamin/typst.ts/dist/esm/main.bundle.js
/*
THIS IS A GENERATED/BUNDLED FILE BY VITE
if you want to view the source, please visit the github repository https://github.com/Myriad-Dreamin/typst.ts/blob/main/packages/typst.ts
*/
var je = Object.defineProperty;
var Be = (r, t, e) => t in r ? je(r, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : r[t] = e;
var p = (r, t, e) => Be(r, typeof t != "symbol" ? t + "" : t, e);
class ze {
  constructor() {
    p(this, "loadedFonts", /* @__PURE__ */ new Set());
    p(this, "fetcher", fetch);
  }
  setFetcher(t) {
    this.fetcher = t;
  }
  async loadFonts(t, e) {
    const n = new Function("m", "return import(m)"), i = this.fetcher || (this.fetcher = await (async function() {
      const { fetchBuilder: u, FileSystemCache: d } = await n("node-fetch-cache"), g = new d({
        /// By default, we don't have a complicated cache policy.
        cacheDirectory: ".cache/typst/fonts"
      }), w = u.withCache(g);
      return function(m, C) {
        const R = setTimeout(() => {
          console.warn("font fetching is stucking:", m);
        }, 15e3);
        return w(m, C).finally(() => {
          clearTimeout(R);
        });
      };
    })()), s = e.filter((u) => u instanceof Uint8Array || typeof u == "object" && "info" in u ? !0 : this.loadedFonts.has(u) ? !1 : (this.loadedFonts.add(u), !0)), o = await Promise.all(
      s.map(async (u) => {
        if (u instanceof Uint8Array) {
          await t.add_raw_font(u);
          return;
        }
        if (typeof u == "object" && "info" in u) {
          await t.add_lazy_font(u, "blob" in u ? u.blob : pe(u));
          return;
        }
        return new Uint8Array(await (await i(u)).arrayBuffer());
      })
    );
    for (const u of o)
      u && await t.add_raw_font(u);
  }
  async build(t, e, n) {
    const i = { ref: this, builder: e, hooks: n };
    for (const s of (t == null ? void 0 : t.beforeBuild) ?? [])
      await s(void 0, i);
    return n.latelyBuild && n.latelyBuild(i), await e.build();
  }
}
async function be(r, t, e, n) {
  var i;
  return await t.init((i = r == null ? void 0 : r.getModule) == null ? void 0 : i.call(r)), await new ze().build(r, new e(), n);
}
function pe(r) {
  return () => {
    const t = new XMLHttpRequest();
    return t.overrideMimeType("text/plain; charset=x-user-defined"), t.open("GET", r.url, !1), t.send(null), t.status === 200 && (t.response instanceof String || typeof t.response == "string") ? Uint8Array.from(t.response, (e) => e.charCodeAt(0)) : new Uint8Array();
  };
}
const qe = [
  "DejaVuSansMono-Bold.ttf",
  "DejaVuSansMono-BoldOblique.ttf",
  "DejaVuSansMono-Oblique.ttf",
  "DejaVuSansMono.ttf",
  "LibertinusSerif-Bold.otf",
  "LibertinusSerif-BoldItalic.otf",
  "LibertinusSerif-Italic.otf",
  "LibertinusSerif-Regular.otf",
  "LibertinusSerif-Semibold.otf",
  "LibertinusSerif-SemiboldItalic.otf",
  "NewCM10-Bold.otf",
  "NewCM10-BoldItalic.otf",
  "NewCM10-Italic.otf",
  "NewCM10-Regular.otf",
  "NewCMMath-Bold.otf",
  "NewCMMath-Book.otf",
  "NewCMMath-Regular.otf"
], Ue = [
  "InriaSerif-Bold.ttf",
  "InriaSerif-BoldItalic.ttf",
  "InriaSerif-Italic.ttf",
  "InriaSerif-Regular.ttf",
  "Roboto-Regular.ttf",
  "NotoSerifCJKsc-Regular.otf"
], Ne = ["TwitterColorEmoji.ttf", "NotoColorEmoji-Regular-COLR.subset.ttf"];
function He(r) {
  var e, n;
  const t = [];
  if (r && (r == null ? void 0 : r.assets) !== !1 && ((e = r == null ? void 0 : r.assets) != null && e.length) && ((n = r == null ? void 0 : r.assets) == null ? void 0 : n.length) > 0) {
    let i = {
      text: "https://cdn.jsdelivr.net/gh/typst/typst-assets@v0.13.1/files/fonts/",
      _: "https://cdn.jsdelivr.net/gh/typst/typst-dev-assets@v0.13.1/files/fonts/"
    }, s = r.assetUrlPrefix ?? i;
    typeof s == "string" ? s = { _: s } : s = { ...i, ...s };
    for (const u of Object.keys(s)) {
      const d = s[u];
      d[d.length - 1] !== "/" && (s[u] = d + "/");
    }
    const o = (u, d) => d.map((g) => (s[u] || s._) + g);
    for (const u of r.assets)
      switch (u) {
        case "text":
          t.push(...o(u, qe));
          break;
        case "cjk":
          t.push(...o(u, Ue));
          break;
        case "emoji":
          t.push(...o(u, Ne));
          break;
      }
  }
  return t;
}
function Nr(r, t) {
  return Z(r, t);
}
function Z(r, t) {
  const e = He(t), n = async (i, { ref: s, builder: o }) => {
    t != null && t.fetcher && s.setFetcher(t.fetcher), await s.loadFonts(o, [...r, ...e]);
  };
  return n._preloadRemoteFontOptions = t, n._kind = "fontLoader", n;
}
function qt({ byFamily: r }) {
  return async (t, { builder: e }) => {
    const n = performance.now();
    if ("queryLocalFonts" in window) {
      const s = await window.queryLocalFonts();
      r = r ?? [];
      for (const o of s) {
        if (!r.includes(o.family))
          continue;
        const u = await (await o.blob()).arrayBuffer();
        await e.add_raw_font(new Uint8Array(u));
      }
    }
    const i = performance.now();
    console.log("preload system font time used:", i - n);
  };
}
function Je(r) {
  return async (t, { builder: e }) => new Promise((n) => {
    e.set_package_registry(r, function(i) {
      return r.resolve(i, this);
    }), n();
  });
}
function Ve(r) {
  return async (t, e) => {
    var n;
    if (e.alreadySetAccessModel)
      throw new Error(
        `already set some assess model before: ${(n = e.alreadySetAccessModel.constructor) == null ? void 0 : n.name}(${e.alreadySetAccessModel})`
      );
    return e.alreadySetAccessModel = r, new Promise((i) => {
      e.builder.set_access_model(
        r,
        (s) => {
          const o = r.getMTime(s);
          return o ? o.getTime() : 0;
        },
        (s) => r.isFile(s) || !1,
        (s) => r.getRealPath(s) || s,
        (s) => r.readAll(s)
      ), i();
    });
  };
}
const k = Symbol.for("reflexo-obj");
var Bt = /* @__PURE__ */ ((r) => (r[r.PIXEL_PER_PT = 3] = "PIXEL_PER_PT", r))(Bt || {});
class he {
  constructor(t, e, n) {
    p(this, "loadPageCount");
    p(this, "imageScaleFactor");
    p(this, "container");
    p(this, "canvasList");
    p(this, "textLayerList");
    p(this, "commonList");
    p(this, "textLayerParentList");
    p(this, "semanticLayerList");
    this.pageInfos = t, this.imageScaleFactor = n.pixelPerPt ?? Bt.PIXEL_PER_PT, e.innerHTML = "", e.style.width = "100%", this.container = e, this.canvasList = new Array(this.loadPageCount), this.textLayerList = new Array(this.loadPageCount), this.commonList = new Array(this.loadPageCount), this.textLayerParentList = new Array(this.loadPageCount), this.semanticLayerList = new Array(this.loadPageCount);
    const i = (s, o, u) => {
      const d = Math.ceil(o.width) * this.imageScaleFactor, g = Math.ceil(o.height) * this.imageScaleFactor, w = this.canvasList[s] = document.createElement("canvas"), m = this.semanticLayerList[s] = document.createElement("div"), C = this.textLayerList[s] = document.createElement("div"), R = this.textLayerParentList[s] = document.createElement("div");
      if (w.getContext("2d")) {
        const $ = document.createElement("div");
        w.width = d, w.height = g, $.appendChild(w), u.appendChild($), $.style.position = "absolute";
      }
      {
        R.appendChild(C), R.className = "typst-html-semantics";
        const $ = e.offsetWidth, B = $ / o.width;
        R.style.width = `${$}px`, R.style.height = `${o.height * B}px`, R.style.setProperty("--data-text-width", `${B}px`), R.style.setProperty("--data-text-height", `${B}px`), u.classList.add("typst-page"), u.classList.add("canvas"), u.style.width = `${$}px`, u.style.height = `${g * B}px`, u.style.position = "relative", m.appendChild(R), u.appendChild(m);
      }
    };
    for (let s = 0; s < this.pageInfos.length; s++) {
      const o = this.pageInfos[s];
      let u;
      u = this.commonList[s] = document.createElement("div"), e.appendChild(u), i(s, o, u);
    }
  }
  resetLayout() {
    for (let t = 0; t < this.pageInfos.length; t++) {
      const e = this.pageInfos[t], n = Math.ceil(e.width) * this.imageScaleFactor, i = Math.ceil(e.height) * this.imageScaleFactor, s = this.canvasList[t].parentElement;
      if (!s)
        throw new Error(
          `canvasDiv is null for page ${t}, canvas list length ${this.canvasList.length}`
        );
      const o = this.commonList[t], u = this.textLayerParentList[t], d = this.container.offsetWidth, g = d / n;
      u.style.width = `${d}px`, u.style.height = `${i * g}px`, o.style.width = `${d}px`, o.style.height = `${i * g}px`;
      const w = this.container.offsetWidth / n;
      s.style.transformOrigin = "0px 0px", s.style.transform = `scale(${w})`;
    }
  }
}
const Ge = (r) => {
  let t = !1, e;
  return () => t ? e : (t = !0, e = r());
};
class me {
  constructor(t) {
    p(this, "wasmBin");
    p(this, "initOnce");
    if (typeof t != "function")
      throw new Error("initFn is not a function");
    this.initOnce = Ge(async () => {
      await t(this.wasmBin);
    });
  }
  async init(t) {
    this.wasmBin = t, await this.initOnce();
  }
}
class zt {
  constructor(t) {
    p(this, "hookedElem");
    p(this, "kModule");
    p(this, "opts");
    p(this, "modes", []);
    /// Configuration fields
    /// enable partial rendering
    p(this, "partialRendering", !0);
    /// underlying renderer
    p(this, "renderMode", "svg");
    p(this, "r");
    /// preview mode
    p(this, "previewMode", 0);
    /// whether this is a content preview
    p(this, "isContentPreview", !1);
    /// whether this content preview will mix outline titles
    p(this, "isMixinOutline", !1);
    /// background color
    p(this, "backgroundColor", "black");
    /// default page color (empty string means transparent)
    p(this, "pageColor", "white");
    /// pixel per pt
    p(this, "pixelPerPt", 3);
    /// customized way to retrieving dom state
    p(this, "retrieveDOMState");
    /// State fields
    /// whether svg is updating (in triggerSvgUpdate)
    p(this, "isRendering", !1);
    /// whether kModule is initialized
    p(this, "moduleInitialized", !1);
    /// patch queue for updating data.
    p(this, "patchQueue", []);
    /// resources to dispose
    p(this, "disposeList", []);
    /// canvas render ctoken
    p(this, "canvasRenderCToken");
    /// There are two scales in this class: The real scale is to adjust the size
    /// of `hookedElem` to fit the svg. The virtual scale (scale ratio) is to let
    /// user zoom in/out the svg. For example:
    /// + the default value of virtual scale is 1, which means the svg is totally
    ///   fit in `hookedElem`.
    /// + if user set virtual scale to 0.5, then the svg will be zoomed out to fit
    ///   in half width of `hookedElem`. "real" current scale of `hookedElem`
    p(this, "currentRealScale", 1);
    /// "virtual" current scale of `hookedElem`
    p(this, "currentScaleRatio", 1);
    /// timeout for delayed viewport change
    p(this, "vpTimeout");
    /// sampled by last render time.
    p(this, "sampledRenderTime", 0);
    /// page to partial render
    p(this, "partialRenderPage", 0);
    /// outline data
    p(this, "outline");
    /// cursor position in form of [page, x, y]
    p(this, "cursorPosition");
    // id: number = rnd++;
    /// Cache fields
    /// cached state of container, default to retrieve state from `this.hookedElem`
    p(this, "cachedDOMState", {
      width: 0,
      height: 0,
      window: {
        innerWidth: 0,
        innerHeight: 0
      },
      boundingRect: {
        left: 0,
        top: 0,
        right: 0
      }
    });
    var e, n;
    this.hookedElem = t.hookedElem, this.kModule = t.kModule, this.opts = t || {};
    {
      const { renderMode: i, previewMode: s, isContentPreview: o, retrieveDOMState: u } = t || {};
      this.partialRendering = !1, this.renderMode = i ?? this.renderMode, this.previewMode = s ?? this.previewMode, this.isContentPreview = o || !1, this.retrieveDOMState = u ?? (() => ({
        width: this.hookedElem.offsetWidth,
        height: this.hookedElem.offsetHeight,
        window: {
          innerWidth: window.innerWidth,
          innerHeight: window.innerHeight
        },
        boundingRect: this.hookedElem.getBoundingClientRect()
      })), this.backgroundColor = getComputedStyle(document.documentElement).getPropertyValue(
        "--typst-preview-background-color"
      );
    }
    this.hookedElem.classList.add("hide-scrollbar-x"), (e = this.hookedElem.parentElement) == null || e.classList.add("hide-scrollbar-x"), this.previewMode === 1 && (this.hookedElem.classList.add("hide-scrollbar-y"), (n = this.hookedElem.parentElement) == null || n.classList.add("hide-scrollbar-y")), this.installCtrlWheelHandler();
  }
  reset() {
    this.kModule.reset(), this.moduleInitialized = !1;
  }
  dispose() {
    const t = this.disposeList;
    this.disposeList = [], t.forEach((e) => e());
  }
  static derive(t, e) {
    return ["rescale", "rerender", "postRender"].reduce((n, i) => (n[i] = t[`${i}$${e}`].bind(t), console.assert(n[i] !== void 0, `${i}$${e} is undefined`), n), {});
  }
  registerMode(t) {
    const e = zt.derive(this, t);
    this.modes.push([t, e]), t === this.renderMode && (this.r = e);
  }
  installCtrlWheelHandler() {
    const t = [
      0.1,
      0.2,
      0.3,
      0.4,
      0.5,
      0.6,
      0.7,
      0.8,
      0.9,
      1,
      1.1,
      1.3,
      1.5,
      1.7,
      1.9,
      2.1,
      2.4,
      2.7,
      3,
      3.3,
      3.7,
      4.1,
      4.6,
      5.1,
      5.7,
      6.3,
      7,
      7.7,
      8.5,
      9.4,
      10
    ], e = (n) => {
      var i, s, o, u;
      if (n.ctrlKey) {
        n.preventDefault(), this.cachedDOMState = this.retrieveDOMState(), window.onresize !== null && (window.onresize = null);
        const d = this.currentScaleRatio;
        if (n.deltaY < 0) {
          if (this.currentScaleRatio >= t.at(-1))
            return;
          this.currentScaleRatio = t.filter((R) => R > this.currentScaleRatio).at(0);
        } else if (n.deltaY > 0) {
          if (this.currentScaleRatio <= t.at(0))
            return;
          this.currentScaleRatio = t.filter((R) => R < this.currentScaleRatio).at(-1);
        } else
          return;
        const g = this.currentScaleRatio / d, w = n.pageX * (g - 1), m = n.pageY * (g - 1);
        Math.abs(this.currentScaleRatio - 1) < 1e-5 ? (this.hookedElem.classList.add("hide-scrollbar-x"), (i = this.hookedElem.parentElement) == null || i.classList.add("hide-scrollbar-x"), this.previewMode === 1 && (this.hookedElem.classList.add("hide-scrollbar-y"), (s = this.hookedElem.parentElement) == null || s.classList.add("hide-scrollbar-y"))) : (this.hookedElem.classList.remove("hide-scrollbar-x"), (o = this.hookedElem.parentElement) == null || o.classList.remove("hide-scrollbar-x"), this.previewMode === 1 && (this.hookedElem.classList.remove("hide-scrollbar-y"), (u = this.hookedElem.parentElement) == null || u.classList.remove("hide-scrollbar-y")));
        const C = this.hookedElem.firstElementChild;
        if (C) {
          const R = this.getSvgScaleRatio(), L = Number.parseFloat(C.getAttribute("data-height")), $ = Math.ceil(L * R);
          this.hookedElem.style.height = `${$ * 2}px`;
        }
        return window.scrollBy(w, m), this.addViewportChange(), !1;
      }
    };
    this.renderMode !== "dom" && (typeof acquireVsCodeApi < "u" ? (window.addEventListener("wheel", e, {
      passive: !1
    }), this.disposeList.push(() => {
      window.removeEventListener("wheel", e);
    })) : (document.body.addEventListener("wheel", e, {
      passive: !1
    }), this.disposeList.push(() => {
      document.body.removeEventListener("wheel", e);
    })));
  }
  /// Get current scale from html to svg
  // Note: one should retrieve dom state before rescale
  getSvgScaleRatio() {
    const t = this.hookedElem.firstElementChild;
    if (!t)
      return 0;
    const e = this.cachedDOMState, n = Number.parseFloat(
      t.getAttribute("data-width") || t.getAttribute("width") || "1"
    ), i = Number.parseFloat(
      t.getAttribute("data-height") || t.getAttribute("height") || "1"
    );
    return this.currentRealScale = this.previewMode === 1 ? Math.min(e.width / n, e.height / i) : e.width / n, this.currentRealScale * this.currentScaleRatio;
  }
  processQueue(t) {
    const e = t[0];
    switch (e) {
      case "new":
      case "diff-v1":
        return e === "new" && this.reset(), this.kModule.manipulateData({
          action: "merge",
          data: t[1]
        }), this.moduleInitialized = !0, !0;
      case "viewport-change":
        return this.moduleInitialized ? !0 : (console.log("viewport-change before initialization"), !1);
      default:
        return console.log("svgUpdateEvent", t), !1;
    }
  }
  triggerUpdate() {
    if (this.isRendering)
      return;
    this.isRendering = !0;
    const t = async () => {
      if (this.cachedDOMState = this.retrieveDOMState(), this.patchQueue.length === 0) {
        this.isRendering = !1, this.postprocessChanges();
        return;
      }
      try {
        let e = performance.now();
        const n = this.canvasRenderCToken;
        n && (await n.cancel(), await n.wait(), this.canvasRenderCToken = void 0, console.log("cancel canvas rendering"));
        let i = !1;
        for (; this.patchQueue.length > 0; )
          i = this.processQueue(this.patchQueue.shift()) || i;
        let s = performance.now();
        i && (this.r.rescale(), await this.r.rerender(), this.r.rescale());
        let o = performance.now();
        const u = (d, g, w) => `${d} ${(w - g).toFixed(2)} ms`;
        this.sampledRenderTime = o - e, requestAnimationFrame(t);
      } catch (e) {
        console.error(e), this.isRendering = !1, this.postprocessChanges();
      }
    };
    requestAnimationFrame(t);
  }
  postprocessChanges() {
    this.r.postRender(), this.previewMode === 1 && document.querySelectorAll(".typst-page-number-indicator").forEach((t) => {
      t.textContent = `${this.kModule.retrievePagesInfo().length}`;
    });
  }
  addChangement(t) {
    t[0] === "new" && this.patchQueue.splice(0, this.patchQueue.length);
    const e = () => {
      this.vpTimeout = void 0, this.patchQueue.push(t), this.triggerUpdate();
    };
    this.vpTimeout !== void 0 && clearTimeout(this.vpTimeout), t[0] === "viewport-change" && this.isRendering ? this.vpTimeout = setTimeout(e, this.sampledRenderTime || 100) : e();
  }
  addViewportChange() {
    this.addChangement(["viewport-change", ""]);
  }
}
function Xe(r) {
  return class {
    constructor(e) {
      p(this, "impl");
      p(this, "kModule");
      if (e.isContentPreview && (e.renderMode = "canvas"), this.kModule = e.kModule, this.impl = new r(e), !this.impl.r)
        throw new Error(`mode is not supported, ${e == null ? void 0 : e.renderMode}`);
      e.isContentPreview && (this.impl.partialRendering = !0, this.impl.pixelPerPt = 1, this.impl.isMixinOutline = !0);
    }
    dispose() {
      this.impl.dispose();
    }
    reset() {
      this.impl.reset();
    }
    addChangement(e) {
      this.impl.addChangement(e);
    }
    addViewportChange() {
      this.impl.addViewportChange();
    }
    setPageColor(e) {
      this.impl.pageColor = e, this.addViewportChange();
    }
    setPartialRendering(e) {
      this.impl.partialRendering = e;
    }
    setCursor(e, n, i) {
      this.impl.cursorPosition = [e, n, i];
    }
    setPartialPageNumber(e) {
      return e <= 0 || e > this.kModule.retrievePagesInfo().length ? !1 : (this.impl.partialRenderPage = e - 1, this.addViewportChange(), !0);
    }
    getPartialPageNumber() {
      return this.impl.partialRenderPage + 1;
    }
    setOutineData(e) {
      this.impl.outline = e, this.addViewportChange();
    }
  };
}
function Ke(r, ...t) {
  return t.reduce((e, n) => n(e), r);
}
class Qe {
  constructor() {
    p(this, "isCancellationRequested", !1);
    p(this, "_onCancelled");
    p(this, "_onCancelledResolveResolved");
    let t, e;
    this._onCancelled = new Promise((n) => {
      t = n, e && e(n);
    }), this._onCancelledResolveResolved = new Promise((n) => {
      e = n, t && n(t);
    });
  }
  async cancel() {
    await this._onCancelledResolveResolved, this.isCancellationRequested = !0;
  }
  isCancelRequested() {
    return this.isCancellationRequested;
  }
  async consume() {
    (await this._onCancelledResolveResolved)();
  }
  wait() {
    return this._onCancelled;
  }
}
const Ye = () => new Promise((r) => requestAnimationFrame(r));
function Ze(r) {
  return class extends r {
    constructor(...n) {
      super(...n);
      /// The template element for creating DOM by string.
      p(this, "tmpl", document.createElement("template"));
      /// The stub element for replacing an invisible element.
      p(this, "stub", this.createElement("<stub></stub>"));
      /// Typescript side of lib.
      p(this, "plugin");
      /// Rust side of kernel.
      p(this, "docKernel");
      /// The element to track.
      p(this, "resourceHeader");
      /// Expected exact state of the current DOM.
      /// Initially it is empty meaning no any page is rendered.
      p(this, "pages", []);
      /// The virtual scale of the document.
      p(this, "domScale", 1);
      /// Track mode.
      p(this, "track_mode", 0);
      /// Current executing task.
      p(this, "current_task");
      /// The currently maintained viewport.
      p(this, "viewport");
      if (this.registerMode("dom"), this.disposeList.push(() => {
        this.dispose();
      }), this.plugin = this.opts.renderer, this.opts.domScale !== void 0) {
        if (this.opts.domScale <= 0)
          throw new Error("domScale must be positive");
        this.domScale = this.opts.domScale;
      }
    }
    dispose() {
      for (const n of this.pages)
        n.dispose();
      this.docKernel && this.docKernel.free();
    }
    createElement(n) {
      return this.tmpl.innerHTML = n, this.tmpl.content.firstElementChild;
    }
    async mountDom(n) {
      if (this.docKernel)
        throw new Error("already mounted");
      this.hookedElem.innerHTML = '<svg class="typst-svg-resources" viewBox="0 0 0 0" width="0" height="0" style="opacity: 0; position: absolute;"></svg>', this.resourceHeader = this.hookedElem.querySelector(".typst-svg-resources"), this.docKernel = await this.plugin.renderer.mount_dom(this.kModule[k], this.hookedElem), this.docKernel.bind_functions({
        populateGlyphs: (i) => {
          let o = this.createElement(i).firstElementChild;
          this.resourceHeader.append(o);
        }
      });
    }
    async cancelAnyway$dom() {
      if (this.current_task) {
        const n = this.current_task;
        this.current_task = void 0, await n.cancel();
      }
    }
    retrieveDOMPages() {
      return Array.from(this.hookedElem.querySelectorAll(".typst-dom-page"));
    }
    // doesn't need to postRender
    postRender$dom() {
    }
    // doesn't need to rescale
    rescale$dom() {
    }
    getDomViewport(n, i) {
      const s = i.left, o = -i.top, u = i.right, d = n.innerHeight - i.top, g = {
        x: 0,
        y: o / this.domScale,
        width: Math.max(u - s, 0) / this.domScale,
        height: Math.max(d - o, 0) / this.domScale
      };
      return (g.width <= 0 || g.height <= 0) && (g.x = g.y = g.width = g.height = 0), g;
    }
    // fast mode
    async rerender$dom() {
      const n = this.retrieveDOMState(), { x: i, y: s, width: o, height: u } = this.getDomViewport(n.window, n.boundingRect);
      if (!await this.docKernel.relayout(i, s, o, u))
        return;
      const g = new Qe();
      this.doRender$dom(g), this.current_task = g;
    }
    async doRender$dom(n) {
      const i = (d, g) => {
        if (d && !n.isCancelRequested() && g)
          return g();
      }, s = this.retrieveDOMPages().map((d) => {
        const { innerWidth: g, innerHeight: w } = window, m = d.getBoundingClientRect();
        return {
          inWindow: !(m.left > g || m.right < 0 || m.top > w || m.bottom < 0),
          page: d
        };
      }), o = async (d) => {
        if (await Ye(), n.isCancelRequested())
          return;
        const g = s[d].page, w = g.getBoundingClientRect(), m = this.getDomViewport(window, w), C = (z) => this.docKernel.need_repaint(d, m.x, m.y, m.width, m.height, z), R = (z) => this.docKernel.repaint(d, m.x, m.y, m.width, m.height, z), L = (z) => {
          if (!n.isCancelRequested())
            return i(C(z), () => R(z));
        };
        await L(
          0
          /* Layout */
        );
        const $ = (w.width ? Number.parseFloat(g.getAttribute("data-width")) / w.width : 1) * this.domScale, B = (w.height ? Number.parseFloat(g.getAttribute("data-height")) / w.height : 1) * this.domScale;
        m.x *= $, m.y *= B, m.y -= 100, m.width *= $, m.height *= B, m.height += 200, await L(
          1
          /* Svg */
        ), await L(
          2
          /* Semantics */
        ), !n.isCancelRequested() && (C(
          3
          /* PrepareCanvas */
        ) ? (async () => {
          if (await R(
            3
            /* PrepareCanvas */
          ), !n.isCancelRequested())
            return L(
              4
              /* Canvas */
            );
        })() : await L(
          4
          /* Canvas */
        ));
      }, u = async (d) => {
        for (let g = 0; g < s.length; ++g) {
          if (n.isCancelRequested())
            return;
          s[g].inWindow === d && await o(g);
        }
      };
      this.cancelAnyway$dom(), await u(!0), await u(!1), n.isCancelRequested();
    }
  };
}
class tr extends Xe(
  Ke(
    zt,
    Ze
  )
) {
}
var ge;
let Et = (ge = k, class {
  /**
   * @internal
   */
  constructor(t, e) {
    /**
     * @internal
     */
    p(this, ge);
    this.plugin = t, this[k] = e;
  }
  /**
   * @deprecated set in {@link RenderToCanvasOptions} instead
   *
   * Set the background color of the Typst document.
   * @param {string} t - The background color in format of `^#?[0-9a-f]{6}$`
   *
   * Note: Default to `#ffffff`.
   *
   * Note: Only available in canvas rendering mode.
   */
  set backgroundColor(t) {
    t !== void 0 && (this[k].background_color = t);
  }
  /**
   * Get the background color of the Typst document.
   *
   * Note: Default to `#ffffff`.
   *
   * Note: Only available in canvas rendering mode.
   */
  get backgroundColor() {
    return this[k].background_color;
  }
  /**
   * Set the pixel per point scale up the canvas panel.
   *
   * Note: Default to `3`.
   *
   * Note: Only available in canvas rendering mode.
   */
  set pixelPerPt(t) {
    t !== void 0 && (this[k].pixel_per_pt = t);
  }
  /**
   * @deprecated set in {@link RenderToCanvasOptions} instead
   *
   * Get the pixel per point scale up the canvas panel.
   *
   * Note: Default to `3`.
   *
   * Note: Only available in canvas rendering mode.
   */
  get pixelPerPt() {
    return this[k].pixel_per_pt;
  }
  /**
   * Reset state
   */
  reset() {
    this.plugin.resetSession(this);
  }
  /**
   * @deprecated
   * use {@link docWidth} instead
   */
  get doc_width() {
    return this[k].doc_width;
  }
  get docWidth() {
    return this[k].doc_width;
  }
  /**
   * @deprecated
   * use {@link docHeight} instead
   */
  get doc_height() {
    return this[k].doc_height;
  }
  get docHeight() {
    return this[k].doc_height;
  }
  retrievePagesInfo() {
    const t = this[k].pages_info, e = [], n = t.page_count;
    for (let i = 0; i < n; i++) {
      const s = t.page(i);
      e.push({
        pageOffset: s.page_off,
        width: s.width_pt,
        height: s.height_pt
      });
    }
    return e;
  }
  getSourceLoc(t) {
    return this[k].source_span(t);
  }
  /**
   * See {@link TypstRenderer#renderSvg} for more details.
   */
  renderSvg(t) {
    return this.plugin.renderSvg({
      renderSession: this,
      ...t
    });
  }
  /**
   * See {@link TypstRenderer#renderToSvg} for more details.
   */
  renderToSvg(t) {
    return this.plugin.renderToSvg({
      renderSession: this,
      ...t
    });
  }
  /**
   * See {@link TypstRenderer#renderCanvas} for more details.
   */
  renderCanvas(t) {
    return this.plugin.renderCanvas({
      renderSession: this,
      ...t
    });
  }
  /**
   * See {@link TypstRenderer#manipulateData} for more details.
   */
  manipulateData(t) {
    this.plugin.manipulateData({
      renderSession: this,
      ...t
    });
  }
  /**
   * See {@link TypstRenderer#renderSvgDiff} for more details.
   */
  renderSvgDiff(t) {
    return this.plugin.renderSvgDiff({
      renderSession: this,
      ...t
    });
  }
  /**
   * @deprecated
   * use {@link getSourceLoc} instead
   */
  get_source_loc(t) {
    return this[k].source_span(t);
  }
  /**
   * @deprecated
   * use {@link renderSvgDiff} instead
   */
  render_in_window(t, e, n, i) {
    return this[k].render_in_window(t, e, n, i);
  }
  /**
   * @deprecated
   * use {@link manipulateData} instead
   */
  merge_delta(t) {
    this.plugin.manipulateData({
      renderSession: this,
      action: "merge",
      data: t
    });
  }
});
var fe;
let er = (fe = k, class {
  /**
   * @internal
   */
  constructor(t, e) {
    /**
     * @internal
     */
    p(this, fe);
    /**
     * @internal
     */
    p(this, "managedCanvasElemList", /* @__PURE__ */ new Map());
    /**
     * @internal
     */
    p(this, "canvasCounter", Math.random());
    this.plugin = t, this[k] = e;
  }
  /**
   * See {@link TypstRenderer#manipulateData} for more details.
   */
  manipulateData(t, e) {
    return this[k].manipulate_data(t, e);
  }
  /**
   * You must submit all canvas in pages to ensure synchronization with the background worker
   *
   * See {@link TypstRenderer#renderCanvas} for more details.
   */
  renderCanvas(t) {
    const e = this.managedCanvasElemList;
    for (const [u, d] of e)
      d[0] = 0;
    for (const u of t) {
      const d = u.canvas;
      let g = d.dataset.manageId, w = 2;
      g || (g = this.canvasCounter.toFixed(5), this.canvasCounter += 1, d.dataset.manageId = g, w = 1);
      let m = e.get(g);
      if (m && m[0] !== 0)
        throw new Error("cannot update a canvas for two times in batch");
      e.set(g, [w, { ...u }]);
    }
    const n = Array.from(e.entries()), i = new Uint8Array(n.length), s = new Array(n.length), o = n.map(([u, [d, g]], w) => (d || e.delete(u), i[w] = d, s[w] = g.canvas, this.plugin.canvasOptionsToRust(g)));
    return this[k].render_canvas(i, s, o);
  }
  async retrievePagesInfo() {
    const t = await this[k].get_pages_info();
    console.log(t);
    const e = [], n = t.page_count;
    for (let i = 0; i < n; i++) {
      const s = t.page(i);
      e.push({
        pageOffset: s.page_off,
        width: s.width_pt,
        height: s.height_pt
      });
    }
    return e;
  }
});
const rr = (r) => new me(async (t) => await r.default(t));
function Ut() {
  return new nr();
}
async function Xr() {
  return (await Promise.resolve().then(() => Re)).renderer_build_info();
}
let Nt = !0;
class nr {
  constructor() {
    p(this, "renderer");
    p(this, "rendererJs");
  }
  async init(t) {
    var n;
    this.rendererJs = await (((n = t == null ? void 0 : t.getWrapper) == null ? void 0 : n.call(t)) || Promise.resolve().then(() => Re));
    const e = this.rendererJs.TypstRendererBuilder;
    this.renderer = await be(
      t,
      rr(this.rendererJs),
      e,
      {}
    );
  }
  loadGlyphPack(t) {
    return Promise.resolve();
  }
  createOptionsToRust(t) {
    const e = new this.rendererJs.CreateSessionOptions();
    return t.format !== void 0 && (e.format = t.format), t.artifactContent !== void 0 && (e.artifact_content = t.artifactContent), e;
  }
  canvasOptionsToRust(t) {
    const e = new this.rendererJs.RenderPageImageOptions();
    if (t.pageOffset === void 0)
      throw new Error("pageOffset is required in reflexo v0.5.0");
    if (e.page_off = t.pageOffset, t.cacheKey !== void 0 && (e.cache_key = t.cacheKey), t.backgroundColor !== void 0 && (e.background_color = t.backgroundColor), t.pixelPerPt !== void 0 && (e.pixel_per_pt = t.pixelPerPt), t.dataSelection !== void 0) {
      let n = 0;
      t.dataSelection.body ? n |= 1 : t.canvas && Nt && (Nt = !1, console.warn("dataSelection.body is not set but providing canvas for body")), (t.dataSelection.text || t.dataSelection.annotation) && console.error("dataSelection.text and dataSelection.annotation are deprecated"), t.dataSelection.semantics && (n |= 8), e.data_selection = n;
    }
    return e;
  }
  retrievePagesInfoFromSession(t) {
    return t.retrievePagesInfo();
  }
  /**
   * Render a Typst document to canvas.
   */
  renderCanvas(t) {
    return this.withinOptionSession(t, async (e) => this.renderer.render_page_to_canvas(
      e[k],
      t.canvas || void 0,
      this.canvasOptionsToRust(t)
    ));
  }
  // async renderPdf(artifactContent: string): Promise<Uint8Array> {
  // return this.renderer.render_to_pdf(artifactContent);
  // }
  async inAnimationFrame(t) {
    return new Promise((e, n) => {
      requestAnimationFrame(() => {
        try {
          e(t());
        } catch (i) {
          n(i);
        }
      });
    });
  }
  async renderDisplayLayer(t, e, n) {
    const s = t[k].pages_info.page_count, o = async (w, m) => {
      const R = e[w].getContext("2d");
      if (!R)
        throw new Error("canvas context is null");
      return await this.renderCanvas({
        ...n,
        canvas: R,
        renderSession: t,
        pageOffset: m
      });
    }, u = performance.now(), d = await (async () => {
      const w = [];
      for (let m = 0; m < s; m++)
        w.push(await this.inAnimationFrame(() => o(m, m)));
      return w;
    })(), g = performance.now();
    return console.log(`display layer used: render = ${(g - u).toFixed(1)}ms`), d;
  }
  renderTextLayer(t, e) {
    const n = performance.now();
    t.forEach((s, o) => {
      s.innerHTML = e[o].htmlSemantics[0];
    });
    const i = performance.now();
    console.log(`text layer used: render = ${(i - n).toFixed(1)}ms`);
  }
  async render(t) {
    if ("format" in t && t.format !== "vector" && ["serde_json", "js", "ir"].includes(t.format))
      throw new Error(`deprecated format ${t.format}, please use vector format`);
    return this.renderToCanvas(t);
  }
  async renderDom(t) {
    if ("format" in t && t.format !== "vector" && ["serde_json", "js", "ir"].includes(t.format))
      throw new Error(`deprecated format ${t.format}, please use vector format`);
    return this.withinOptionSession(t, async (e) => {
      const n = new tr({
        ...t,
        renderMode: "dom",
        hookedElem: t.container,
        kModule: e,
        renderer: this
      });
      return await n.impl.mountDom(t.pixelPerPt), n;
    });
  }
  async renderToCanvas(t) {
    let e, n;
    const i = t.container;
    i.style.visibility = "hidden";
    const s = async (o, u) => {
      try {
        n = await this.renderDisplayLayer(e, o, t), u();
      } finally {
        i.style.visibility = "visible";
      }
    };
    return this.withinOptionSession(t, async (o) => {
      if (e = o, e[k].pages_info.page_count === 0)
        throw new Error("No page found in session");
      if (t.pixelPerPt !== void 0 && t.pixelPerPt <= 0)
        throw new Error(
          "Invalid typst.RenderOptions.pixelPerPt, should be a positive number " + t.pixelPerPt
        );
      let u = t.backgroundColor;
      if (u !== void 0 && !/^#[0-9a-f]{6}$/.test(u))
        throw new Error(
          "Invalid typst.backgroundColor color for matching ^#?[0-9a-f]{6}$ " + u
        );
      e.pixelPerPt = t.pixelPerPt ?? Bt.PIXEL_PER_PT, e.backgroundColor = u ?? "#ffffff";
      const d = performance.now(), g = new he(
        this.retrievePagesInfoFromSession(e),
        i,
        t
      ), w = performance.now();
      console.log(`layer used: retrieve = ${(w - d).toFixed(1)}ms`), await s(g.canvasList, () => g.resetLayout()), this.renderTextLayer(g.textLayerList, n);
    });
  }
  createModule(t) {
    return Promise.resolve(
      new Et(
        this,
        this.renderer.create_session(
          t && this.createOptionsToRust({
            format: "vector",
            artifactContent: t
          })
        )
      )
    );
  }
  async createWorkerV0(t) {
    return new er(this, await this.renderer.create_worker(t));
  }
  workerBridge() {
    return this.renderer.create_worker_bridge();
  }
  renderSvg(t, e) {
    if (t instanceof Et || e)
      throw new Error("removed api, please use renderToSvg({ renderSession, container }) instead");
    return this.withinOptionSession(t, async (n) => {
      let i;
      return t.data_selection && (i = 0, t.data_selection.body && (i |= 1), t.data_selection.defs && (i |= 2), t.data_selection.css && (i |= 4), t.data_selection.js && (i |= 8)), Promise.resolve(this.renderer.svg_data(n[k], i));
    });
  }
  renderSvgDiff(t) {
    return t.window ? this.renderer.render_svg_diff(
      t.renderSession[k],
      t.window.lo.x,
      t.window.lo.y,
      t.window.hi.x,
      t.window.hi.y
    ) : this.renderer.render_svg_diff(
      t.renderSession[k],
      0,
      0,
      1e33,
      1e33
    );
  }
  renderToSvg(t) {
    return this.withinOptionSession(t, async (e) => Promise.resolve(this.renderer.render_svg(e[k], t.container)));
  }
  getCustomV1(t) {
    return Promise.resolve(this.renderer.get_customs(t.renderSession[k]));
  }
  resetSession(t) {
    return this.renderer.reset(t[k]);
  }
  manipulateData(t) {
    return this.renderer.manipulate_data(
      t.renderSession[k],
      t.action ?? "reset",
      t.data
    );
  }
  withinOptionSession(t, e) {
    function n(i) {
      return "artifactContent" in i;
    }
    if ("renderSession" in t)
      return e(t.renderSession);
    if (n(t))
      return this.runWithSession(t, e);
    throw new Error(
      "Invalid render options, should be one of RenderByContentOptions|RenderBySessionOptions"
    );
  }
  async runWithSession(t, e) {
    let n = t, i = e;
    e || (n = void 0, i = t);
    const s = this.renderer.create_session(
      /* moved */
      n && this.createOptionsToRust(n)
    );
    try {
      const o = await i(new Et(this, s));
      return s.free(), o;
    } catch (o) {
      throw s.free(), o;
    }
  }
}
var we;
we = k;
class ir {
  /**
   * @internal
   */
  constructor(t) {
    /**
     * @internal
     */
    p(this, we);
    this[k] = t;
  }
  /**
   * Reset the incremental server to the initial state.
   */
  reset() {
    this[k].reset();
  }
  /**
   * Return current result.
   */
  current() {
    return this[k].current();
  }
  /**
   * Also attach the debug info to the result.
   */
  setAttachDebugInfo(t) {
    this[k].set_attach_debug_info(t);
  }
}
function sr() {
  return new _r();
}
var le;
le = k;
class or {
  constructor(t) {
    p(this, le);
    this[k] = t;
  }
  /**
   * Compile the paged document.
   *
   * @param {DiagnosticsFormat} format - The format of the diagnostics.
   * @returns {Promise<{ diagnostics?: DiagnosticsData[DiagnosticsFormat][] }>} - The result of the compilation.
   */
  compile(t) {
    return this[k].compile(0, tt(t == null ? void 0 : t.diagnostics));
  }
  /**
   * Compile the paged document.
   *
   * @param {DiagnosticsFormat} format - The format of the diagnostics.
   * @returns {Promise<{ diagnostics?: DiagnosticsData[DiagnosticsFormat][] }>} - The result of the compilation.
   */
  compileHtml(t) {
    return this[k].compile(1, tt(t == null ? void 0 : t.diagnostics));
  }
  /**
   * Runs query on the paged document.
   */
  async query(t) {
    return JSON.parse(this[k].query(0, t.selector, t.field));
  }
  /**
   * Get the title of the paged document.
   * Throw error if the world didn't compile the paged document.
   *
   * @returns {string | undefined} - The title of the paged document.
   */
  title() {
    return this[k].title(0);
  }
  /**
   * Export the paged document as vector format.
   *
   * @returns {Uint8Array | undefined} - The title of the paged document.
   */
  vector(t) {
    return this[k].get_artifact(0, tt(t == null ? void 0 : t.diagnostics)) || {};
  }
  /**
   * Export the paged document to PDF.
   *
   * @returns {Uint8Array | undefined} - The title of the paged document.
   */
  pdf(t) {
    return this[k].get_artifact(1, tt(t == null ? void 0 : t.diagnostics)) || {};
  }
}
const ye = (r) => new me(async (t) => await r.default(t));
function ve() {
  return new vt();
}
class _r {
  constructor() {
    p(this, "fontBuilderJs");
    p(this, "fontBuilder");
  }
  async init(t) {
    var e, n;
    this.fontBuilderJs = await (((e = t == null ? void 0 : t.getWrapper) == null ? void 0 : e.call(t)) || Promise.resolve().then(() => Oe)), await ye(this.fontBuilderJs).init((n = t == null ? void 0 : t.getModule) == null ? void 0 : n.call(t)), this.fontBuilder = new this.fontBuilderJs.TypstFontResolverBuilder();
  }
  async getFontInfo(t) {
    return this.fontBuilder.get_font_info(t);
  }
  async addFontData(t) {
    this.fontBuilder.add_raw_font(t);
  }
  async addLazyFont(t, e) {
    return this.fontBuilder.add_lazy_font(t, e);
  }
  async build(t) {
    const e = await this.fontBuilder.build(), n = await t(e);
    return e.free(), n;
  }
}
const xt = class xt {
  constructor() {
    p(this, "compiler");
    p(this, "compilerJs");
  }
  async init(t) {
    var g;
    this.compilerJs = await (((g = t == null ? void 0 : t.getWrapper) == null ? void 0 : g.call(t)) || Promise.resolve().then(() => Oe));
    const e = this.compilerJs.TypstCompilerBuilder, n = { ...t || {} }, i = n.beforeBuild ?? (n.beforeBuild = []), s = i.some(
      (w) => w._preloadRemoteFontOptions !== void 0
    ), o = i.some(
      (w) => {
        var m;
        return ((m = w._preloadRemoteFontOptions) == null ? void 0 : m.assets) !== void 0;
      }
    ), u = i.some(
      (w) => {
        var m;
        return ((m = w._preloadRemoteFontOptions) == null ? void 0 : m.assets) === !1;
      }
    );
    if ((!s || !o && !u) && i.push(Z([], { assets: xt.defaultAssets })), !i.some((w) => w._kind === "fontLoader"))
      throw new Error(
        "TypstCompiler: no font loader found, please use font loaders, e.g. loadFonts or preloadSystemFonts"
      );
    this.compiler = await be(t, ye(this.compilerJs), e, {});
  }
  setFonts(t) {
    this.compiler.set_fonts(t);
  }
  compile(t) {
    return new Promise((e) => {
      const n = this.compiler.snapshot(
        t.root,
        t.mainFilePath,
        Ht(t.inputs)
      );
      if ("incrementalServer" in t) {
        e(
          n.incr_compile(
            t.incrementalServer[k],
            tt(t.diagnostics)
          )
        );
        return;
      }
      e(
        n.get_artifact(t.format || 0, tt(t.diagnostics))
      );
    });
  }
  async runWithWorld(t, e) {
    const n = this.compiler.snapshot(
      t.root,
      t.mainFilePath,
      Ht(t.inputs)
    );
    let i = await e(new or(n));
    return n.free(), i;
  }
  query(t) {
    return this.runWithWorld(t, async (e) => JSON.parse(await e.query(t)));
  }
  getSemanticTokenLegend() {
    return new Promise((t) => {
      t(this.compiler.get_semantic_token_legend());
    });
  }
  getSemanticTokens(t) {
    return new Promise((e) => {
      this.compiler.reset(), e(
        this.compiler.get_semantic_tokens(
          t.offsetEncoding || "utf-16",
          t.mainFilePath,
          t.resultId
        )
      );
    });
  }
  async withIncrementalServer(t) {
    const e = new ir(this.compiler.create_incr_server());
    try {
      return await t(e);
    } finally {
      e[k].free();
    }
  }
  async getAst(t) {
    return this.compiler.get_ast(t);
  }
  async reset() {
    await new Promise((t) => {
      this.compiler.reset(), t(void 0);
    });
  }
  addSource(t, e) {
    if (arguments.length > 2)
      throw new Error(
        "use of addSource(path, source, isMain) is deprecated, please use addSource(path, source) instead"
      );
    this.compiler.add_source(t, e);
  }
  mapShadow(t, e) {
    this.compiler.map_shadow(t, e);
  }
  unmapShadow(t) {
    this.compiler.unmap_shadow(t);
  }
  resetShadow() {
    this.compiler.reset_shadow();
  }
  renderPageToCanvas() {
    throw new Error("Please use the api TypstRenderer.renderToCanvas in v0.4.0");
  }
};
p(xt, "defaultAssets", ["text"]);
let vt = xt;
ve._impl = vt;
function Ht(r) {
  return r ? Object.entries(r) : void 0;
}
function tt(r) {
  switch (r) {
    case "none":
      return 1;
    case "unix":
      return 2;
    case "full":
    default:
      return 3;
  }
}
class ar {
  constructor(t, e) {
    p(this, "fullyCached");
    p(this, "mTimes", /* @__PURE__ */ new Map());
    p(this, "mRealPaths", /* @__PURE__ */ new Map());
    p(this, "mData", /* @__PURE__ */ new Map());
    this.root = t, t.endsWith("/") && (this.root = this.root.slice(0, this.root.length - 1)), e != null && e.polyfillHeadRequest, this.fullyCached = !!(e != null && e.fullyCached);
  }
  reset() {
    this.mTimes.clear(), this.mRealPaths.clear(), this.mData.clear();
  }
  resolvePath(t) {
    return this.root + t;
  }
  insertFile(t, e, n) {
    this.mTimes.set(t, n), this.mData.set(t, e);
  }
  removeFile(t) {
    this.mTimes.delete(t), this.mData.delete(t);
  }
  async getPreloadScript() {
    const t = [];
    t.push("((async () => {"), t.push(
      "const snapshot = {  root: '', mTimes: new Map(),  mRealPaths: new Map(),  mData: [],};"
    ), t.push("const runFetch = async (path) => {"), t.push("  const res = await fetch(snapshot.root + path);"), t.push("  const buffer = await res.arrayBuffer();"), t.push("  return [path, new Uint8Array(buffer)];"), t.push("};"), t.push(`snapshot.root = ${JSON.stringify(this.root)};`), t.push(
      `snapshot.mTimes = new Map([${[...this.mTimes.entries()].map(([n, i]) => `[${JSON.stringify(n)}, ${(i == null ? void 0 : i.getTime()) || "undefined"}]`).join(", ")}]);`
    ), t.push(
      `snapshot.mRealPaths = new Map([${[...this.mRealPaths.entries()].map(([n, i]) => `[${JSON.stringify(n)}, ${JSON.stringify(i)}]`).join(", ")}]);`
    );
    const e = await Promise.all(
      [...this.mData.entries()].map(async ([n, i]) => (n = JSON.stringify(n), i ? `runFetch(${n})` : `Promise.resolve([${n}, undefined])`))
    );
    return t.push(`snapshot.mData = await Promise.all([${e.join(", ")}]);`), t.push("return snapshot;"), t.push("})())"), t.join(`
`);
  }
  getLastModified(t) {
    const e = new XMLHttpRequest();
    return e.open("HEAD", t, !1), e.send(null), e.status === 200 ? e.getResponseHeader("Last-Modified") : null;
  }
  getMTimeInternal(t) {
    const e = this.getLastModified(this.resolvePath(t));
    if (e)
      return new Date(e);
  }
  getMTime(t) {
    if (t.startsWith("/@memory/"))
      return this.mTimes.has(t) ? this.mTimes.get(t) : void 0;
    if (!this.fullyCached)
      return this.getMTimeInternal(t);
    if (this.mTimes.has(t))
      return this.mTimes.get(t);
    const e = this.getMTimeInternal(t);
    return this.mTimes.set(t, e), e;
  }
  // todo: isFile
  isFile() {
    return !0;
  }
  // todo: getRealPath
  getRealPath(t) {
    return t;
  }
  readAllInternal(t) {
    const e = new XMLHttpRequest();
    if (e.overrideMimeType("text/plain; charset=x-user-defined"), e.open("GET", this.resolvePath(t), !1), e.send(null), e.status === 200 && (e.response instanceof String || typeof e.response == "string"))
      return Uint8Array.from(e.response, (n) => n.charCodeAt(0));
  }
  readAll(t) {
    if (t.startsWith("/@memory/"))
      return this.mData.has(t) ? this.mData.get(t) : void 0;
    if (!this.fullyCached)
      return this.readAllInternal(t);
    if (this.mData.has(t))
      return this.mData.get(t);
    const e = this.readAllInternal(t);
    return this.mData.set(t, e), e;
  }
}
class cr {
  constructor() {
    p(this, "mTimes", /* @__PURE__ */ new Map());
    p(this, "mData", /* @__PURE__ */ new Map());
  }
  reset() {
    this.mTimes.clear(), this.mData.clear();
  }
  insertFile(t, e, n) {
    this.mTimes.set(t, n), this.mData.set(t, e);
  }
  removeFile(t) {
    this.mTimes.delete(t), this.mData.delete(t);
  }
  getMTime(t) {
    if (t.startsWith("/@memory/") && this.mTimes.has(t))
      return this.mTimes.get(t);
  }
  isFile() {
    return !0;
  }
  getRealPath(t) {
    return t;
  }
  readAll(t) {
    if (t.startsWith("/@memory/") && this.mData.has(t))
      return this.mData.get(t);
  }
}
class ur {
  constructor(t) {
    p(this, "cache", /* @__PURE__ */ new Map());
    this.am = t;
  }
  resolvePath(t) {
    return `https://packages.typst.org/preview/${t.name}-${t.version}.tar.gz`;
  }
  pullPackageData(t) {
    const e = new XMLHttpRequest();
    if (e.overrideMimeType("text/plain; charset=x-user-defined"), e.open("GET", this.resolvePath(t), !1), e.send(null), e.status === 200 && (e.response instanceof String || typeof e.response == "string"))
      return Uint8Array.from(e.response, (n) => n.charCodeAt(0));
  }
  resolve(t, e) {
    if (t.namespace !== "preview")
      return;
    const n = this.resolvePath(t);
    if (this.cache.has(n))
      return this.cache.get(n)();
    const i = this.pullPackageData(t);
    if (!i)
      return;
    const s = `/@memory/fetch/packages/${t.namespace}/${t.name}/${t.version}`, o = [];
    e.untar(i, (d, g, w) => {
      o.push([s + "/" + d, g, new Date(w)]);
    });
    const u = () => {
      for (const [d, g, w] of o)
        this.am.insertFile(d, g, w);
      return s;
    };
    return this.cache.set(n, u), u();
  }
}
typeof window < "u" && (window.TypstRenderModule = {
  RenderView: he,
  createTypstRenderer: Ut,
  createTypstSvgRenderer: Ut,
  preloadRemoteFonts: Z,
  loadFonts: Z,
  preloadSystemFonts: qt
}, window.TypstCompileModule = {
  createTypstCompiler: ve,
  createTypstFontBuilder: sr,
  preloadRemoteFonts: Z,
  loadFonts: Z,
  loadFontSync: pe,
  preloadSystemFonts: qt,
  FetchAccessModel: ar,
  MemoryAccessModel: cr,
  FetchPackageRegistry: ur,
  withAccessModel: Ve,
  withPackageRegistry: Je
});
let _;
function v(r) {
  et === N.length && N.push(N.length + 1);
  const t = et;
  if (et = N[t], typeof et != "number") throw new Error("corrupt heap");
  return N[t] = r, t;
}
function D(r) {
  if (typeof r != "boolean")
    throw new Error(`expected a boolean argument, found ${typeof r}`);
}
function U(r, t) {
  if (!(r instanceof t))
    throw new Error(`expected instance of ${t.name}`);
}
function b(r) {
  if (typeof r != "number") throw new Error(`expected a number argument, found ${typeof r}`);
}
const St = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => r.dtor(r.a, r.b));
function Mt(r) {
  const t = typeof r;
  if (t == "number" || t == "boolean" || r == null)
    return `${r}`;
  if (t == "string")
    return `"${r}"`;
  if (t == "symbol") {
    const i = r.description;
    return i == null ? "Symbol" : `Symbol(${i})`;
  }
  if (t == "function") {
    const i = r.name;
    return typeof i == "string" && i.length > 0 ? `Function(${i})` : "Function";
  }
  if (Array.isArray(r)) {
    const i = r.length;
    let s = "[";
    i > 0 && (s += Mt(r[0]));
    for (let o = 1; o < i; o++)
      s += ", " + Mt(r[o]);
    return s += "]", s;
  }
  const e = /\[object ([^\]]+)\]/.exec(toString.call(r));
  let n;
  if (e && e.length > 1)
    n = e[1];
  else
    return toString.call(r);
  if (n == "Object")
    try {
      return "Object(" + JSON.stringify(r) + ")";
    } catch {
      return "Object";
    }
  return r instanceof Error ? `${r.name}: ${r.message}
${r.stack}` : n;
}
function dr(r) {
  r < 132 || (N[r] = et, et = r);
}
function Jt(r, t) {
  return r = r >>> 0, rt().subarray(r / 1, r / 1 + t);
}
let V = null;
function h() {
  return (V === null || V.buffer.detached === !0 || V.buffer.detached === void 0 && V.buffer !== _.memory.buffer) && (V = new DataView(_.memory.buffer)), V;
}
function E(r, t) {
  return r = r >>> 0, br(r, t);
}
let dt = null;
function gr() {
  return (dt === null || dt.byteLength === 0) && (dt = new Uint32Array(_.memory.buffer)), dt;
}
let gt = null;
function rt() {
  return (gt === null || gt.byteLength === 0) && (gt = new Uint8Array(_.memory.buffer)), gt;
}
function c(r) {
  return N[r];
}
function x(r, t) {
  try {
    return r.apply(this, t);
  } catch (e) {
    _.__wbindgen_export3(v(e));
  }
}
let N = new Array(128).fill(void 0);
N.push(void 0, null, !0, !1);
let et = N.length;
function F(r) {
  return r == null;
}
function f(r, t) {
  try {
    return r.apply(this, t);
  } catch (e) {
    let n = (function() {
      try {
        return e instanceof Error ? `${e.message}

Stack:
${e.stack}` : e.toString();
      } catch {
        return "<failed to stringify thrown value>";
      }
    })();
    throw console.error("wasm-bindgen: imported JS function that was not marked as `catch` threw an error:", n), e;
  }
}
function Vt(r, t, e, n) {
  const i = { a: r, b: t, cnt: 1, dtor: e }, s = (...o) => {
    i.cnt++;
    try {
      return n(i.a, i.b, ...o);
    } finally {
      s._wbg_cb_unref();
    }
  };
  return s._wbg_cb_unref = () => {
    --i.cnt === 0 && (i.dtor(i.a, i.b), i.a = 0, St.unregister(i));
  }, St.register(s, i, i), s;
}
function fr(r, t, e, n) {
  const i = { a: r, b: t, cnt: 1, dtor: e }, s = (...o) => {
    i.cnt++;
    const u = i.a;
    i.a = 0;
    try {
      return n(u, i.b, ...o);
    } finally {
      i.a = u, s._wbg_cb_unref();
    }
  };
  return s._wbg_cb_unref = () => {
    --i.cnt === 0 && (i.dtor(i.a, i.b), i.a = 0, St.unregister(i));
  }, St.register(s, i, i), s;
}
function wr(r, t) {
  const e = t(r.length * 4, 4) >>> 0;
  return gr().set(r, e / 4), P = r.length, e;
}
function It(r, t) {
  const e = t(r.length * 1, 1) >>> 0;
  return rt().set(r, e / 1), P = r.length, e;
}
function Gt(r, t) {
  const e = t(r.length * 4, 4) >>> 0, n = h();
  for (let i = 0; i < r.length; i++)
    n.setUint32(e + 4 * i, v(r[i]), !0);
  return P = r.length, e;
}
function W(r, t, e) {
  if (typeof r != "string") throw new Error(`expected a string argument, found ${typeof r}`);
  if (e === void 0) {
    const u = lt.encode(r), d = t(u.length, 1) >>> 0;
    return rt().subarray(d, d + u.length).set(u), P = u.length, d;
  }
  let n = r.length, i = t(n, 1) >>> 0;
  const s = rt();
  let o = 0;
  for (; o < n; o++) {
    const u = r.charCodeAt(o);
    if (u > 127) break;
    s[i + o] = u;
  }
  if (o !== n) {
    o !== 0 && (r = r.slice(o)), i = e(i, n, n = o + r.length * 3, 1) >>> 0;
    const u = rt().subarray(i + o, i + n), d = lt.encodeInto(r, u);
    if (d.read !== r.length) throw new Error("failed to pass whole string");
    o += d.written, i = e(i, n, o, 1) >>> 0;
  }
  return P = o, i;
}
function M(r) {
  const t = c(r);
  return dr(r), t;
}
let mt = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 });
mt.decode();
const lr = 2146435072;
let Rt = 0;
function br(r, t) {
  return Rt += t, Rt >= lr && (mt = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 }), mt.decode(), Rt = t), mt.decode(rt().subarray(r, r + t));
}
const lt = new TextEncoder();
"encodeInto" in lt || (lt.encodeInto = function(r, t) {
  const e = lt.encode(r);
  return t.set(e), {
    read: r.length,
    written: e.length
  };
});
let P = 0;
function pr(r, t, e) {
  b(r), b(t), _.__wasm_bindgen_func_elem_1056(r, t, v(e));
}
function hr(r, t, e) {
  b(r), b(t), _.__wasm_bindgen_func_elem_1034(r, t, v(e));
}
function mr(r, t) {
  b(r), b(t), _.__wasm_bindgen_func_elem_1035(r, t);
}
function yr(r, t, e, n) {
  b(r), b(t), _.__wasm_bindgen_func_elem_3555(r, t, v(e), v(n));
}
const Xt = ["nonzero", "evenodd"], Kt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_createsessionoptions_free(r >>> 0, 1)), Qt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_incrdomdocclient_free(r >>> 0, 1)), Yt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_pageinfo_free(r >>> 0, 1)), Zt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_pagesinfo_free(r >>> 0, 1)), te = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_renderpageimageoptions_free(r >>> 0, 1)), ee = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_rendersession_free(r >>> 0, 1)), re = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_rendersessionoptions_free(r >>> 0, 1)), At = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_typstrenderer_free(r >>> 0, 1)), ne = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_typstrendererbuilder_free(r >>> 0, 1)), vr = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_typstworker_free(r >>> 0, 1)), ie = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_workerbridge_free(r >>> 0, 1));
class kt {
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, Kt.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_createsessionoptions_free(t, 0);
  }
  constructor() {
    const t = _.createsessionoptions_new();
    return this.__wbg_ptr = t >>> 0, Kt.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @param {string} format
   */
  set format(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const e = W(t, _.__wbindgen_export, _.__wbindgen_export2), n = P;
    _.createsessionoptions_set_format(this.__wbg_ptr, e, n);
  }
  /**
   * @param {Uint8Array} artifact_content
   */
  set artifact_content(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const e = It(t, _.__wbindgen_export), n = P;
    _.createsessionoptions_set_artifact_content(this.__wbg_ptr, e, n);
  }
}
Symbol.dispose && (kt.prototype[Symbol.dispose] = kt.prototype.free);
class it {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(it.prototype);
    return e.__wbg_ptr = t, Qt.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, Qt.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_incrdomdocclient_free(t, 0);
  }
  /**
   * @param {any} functions
   */
  bind_functions(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), _.incrdomdocclient_bind_functions(this.__wbg_ptr, v(t));
  }
  /**
   * Relayout the document in the given window.
   * @param {number} x
   * @param {number} y
   * @param {number} w
   * @param {number} h
   * @returns {Promise<boolean>}
   */
  relayout(t, e, n, i) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const s = _.incrdomdocclient_relayout(this.__wbg_ptr, t, e, n, i);
    return M(s);
  }
  /**
   * @param {number} page_num
   * @param {number} x
   * @param {number} y
   * @param {number} w
   * @param {number} h
   * @param {number} stage
   * @returns {boolean}
   */
  need_repaint(t, e, n, i, s, o) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const w = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), b(t), b(o), _.incrdomdocclient_need_repaint(w, this.__wbg_ptr, t, e, n, i, s, o);
      var u = h().getInt32(w + 0, !0), d = h().getInt32(w + 4, !0), g = h().getInt32(w + 8, !0);
      if (g)
        throw M(d);
      return u !== 0;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} page_num
   * @param {number} x
   * @param {number} y
   * @param {number} w
   * @param {number} h
   * @param {number} stage
   * @returns {any}
   */
  repaint(t, e, n, i, s, o) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const w = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), b(t), b(o), _.incrdomdocclient_repaint(w, this.__wbg_ptr, t, e, n, i, s, o);
      var u = h().getInt32(w + 0, !0), d = h().getInt32(w + 4, !0), g = h().getInt32(w + 8, !0);
      if (g)
        throw M(d);
      return M(u);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (it.prototype[Symbol.dispose] = it.prototype.free);
class X {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(X.prototype);
    return e.__wbg_ptr = t, Yt.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, Yt.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_pageinfo_free(t, 0);
  }
  /**
   * @returns {number}
   */
  get page_off() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.pageinfo_page_off(this.__wbg_ptr) >>> 0;
  }
  /**
   * @returns {number}
   */
  get width_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.pageinfo_width_pt(this.__wbg_ptr);
  }
  /**
   * @returns {number}
   */
  get height_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.pageinfo_height_pt(this.__wbg_ptr);
  }
}
Symbol.dispose && (X.prototype[Symbol.dispose] = X.prototype.free);
class st {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(st.prototype);
    return e.__wbg_ptr = t, Zt.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, Zt.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_pagesinfo_free(t, 0);
  }
  /**
   * @returns {number}
   */
  get page_count() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.pagesinfo_page_count(this.__wbg_ptr) >>> 0;
  }
  /**
   * @param {number} num
   * @returns {PageInfo | undefined}
   */
  page_by_number(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), b(t);
    const e = _.pagesinfo_page_by_number(this.__wbg_ptr, t);
    return e === 0 ? void 0 : X.__wrap(e);
  }
  /**
   * @param {number} i
   * @returns {PageInfo}
   */
  page(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), b(t);
    const e = _.pagesinfo_page(this.__wbg_ptr, t);
    return X.__wrap(e);
  }
  /**
   * @returns {number}
   */
  width() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.pagesinfo_width(this.__wbg_ptr);
  }
  /**
   * @returns {number}
   */
  height() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.pagesinfo_height(this.__wbg_ptr);
  }
}
Symbol.dispose && (st.prototype[Symbol.dispose] = st.prototype.free);
class K {
  static __unwrap(t) {
    return t instanceof K ? t.__destroy_into_raw() : 0;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, te.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_renderpageimageoptions_free(t, 0);
  }
  constructor() {
    const t = _.renderpageimageoptions_new();
    return this.__wbg_ptr = t >>> 0, te.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @returns {number | undefined}
   */
  get pixel_per_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const t = _.renderpageimageoptions_pixel_per_pt(this.__wbg_ptr);
    return t === 4294967297 ? void 0 : t;
  }
  /**
   * @param {number | null} [pixel_per_pt]
   */
  set pixel_per_pt(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), F(t) || b(t), _.renderpageimageoptions_set_pixel_per_pt(this.__wbg_ptr, F(t) ? 4294967297 : Math.fround(t));
  }
  /**
   * @returns {string | undefined}
   */
  get background_color() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), _.renderpageimageoptions_background_color(n, this.__wbg_ptr);
      var t = h().getInt32(n + 0, !0), e = h().getInt32(n + 4, !0);
      let i;
      return t !== 0 && (i = E(t, e).slice(), _.__wbindgen_export4(t, e * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [background_color]
   */
  set background_color(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    var e = F(t) ? 0 : W(t, _.__wbindgen_export, _.__wbindgen_export2), n = P;
    _.renderpageimageoptions_set_background_color(this.__wbg_ptr, e, n);
  }
  /**
   * @returns {number}
   */
  get page_off() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.renderpageimageoptions_page_off(this.__wbg_ptr) >>> 0;
  }
  /**
   * @param {number} page_off
   */
  set page_off(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), b(t), _.renderpageimageoptions_set_page_off(this.__wbg_ptr, t);
  }
  /**
   * @returns {string | undefined}
   */
  get cache_key() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), _.renderpageimageoptions_cache_key(n, this.__wbg_ptr);
      var t = h().getInt32(n + 0, !0), e = h().getInt32(n + 4, !0);
      let i;
      return t !== 0 && (i = E(t, e).slice(), _.__wbindgen_export4(t, e * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [cache_key]
   */
  set cache_key(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    var e = F(t) ? 0 : W(t, _.__wbindgen_export, _.__wbindgen_export2), n = P;
    _.renderpageimageoptions_set_cache_key(this.__wbg_ptr, e, n);
  }
  /**
   * @returns {number | undefined}
   */
  get data_selection() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const t = _.renderpageimageoptions_data_selection(this.__wbg_ptr);
    return t === 4294967297 ? void 0 : t;
  }
  /**
   * @param {number | null} [data_selection]
   */
  set data_selection(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), F(t) || b(t), _.renderpageimageoptions_set_data_selection(this.__wbg_ptr, F(t) ? 4294967297 : t >>> 0);
  }
}
Symbol.dispose && (K.prototype[Symbol.dispose] = K.prototype.free);
class j {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(j.prototype);
    return e.__wbg_ptr = t, ee.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ee.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_rendersession_free(t, 0);
  }
  /**
   * @param {number} rect_lo_x
   * @param {number} rect_lo_y
   * @param {number} rect_hi_x
   * @param {number} rect_hi_y
   * @returns {string}
   */
  render_in_window(t, e, n, i) {
    let s, o;
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const g = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), _.rendersession_render_in_window(g, this.__wbg_ptr, t, e, n, i);
      var u = h().getInt32(g + 0, !0), d = h().getInt32(g + 4, !0);
      return s = u, o = d, E(u, d);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16), _.__wbindgen_export4(s, o, 1);
    }
  }
  /**
   * @returns {number | undefined}
   */
  get pixel_per_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const t = _.rendersession_pixel_per_pt(this.__wbg_ptr);
    return t === 4294967297 ? void 0 : t;
  }
  /**
   * @param {number | null} [pixel_per_pt]
   */
  set pixel_per_pt(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), F(t) || b(t), _.rendersession_set_pixel_per_pt(this.__wbg_ptr, F(t) ? 4294967297 : Math.fround(t));
  }
  /**
   * @returns {string | undefined}
   */
  get background_color() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), _.rendersession_background_color(n, this.__wbg_ptr);
      var t = h().getInt32(n + 0, !0), e = h().getInt32(n + 4, !0);
      let i;
      return t !== 0 && (i = E(t, e).slice(), _.__wbindgen_export4(t, e * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [background_color]
   */
  set background_color(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    var e = F(t) ? 0 : W(t, _.__wbindgen_export, _.__wbindgen_export2), n = P;
    _.rendersession_set_background_color(this.__wbg_ptr, e, n);
  }
  /**
   * @returns {PagesInfo}
   */
  get pages_info() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const t = _.rendersession_pages_info(this.__wbg_ptr);
    return st.__wrap(t);
  }
  /**
   * @returns {number}
   */
  get doc_width() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.rendersession_doc_width(this.__wbg_ptr);
  }
  /**
   * @returns {number}
   */
  get doc_height() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return b(this.__wbg_ptr), _.rendersession_doc_height(this.__wbg_ptr);
  }
  /**
   * @param {Uint32Array} path
   * @returns {string | undefined}
   */
  source_span(t) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr);
      const u = wr(t, _.__wbindgen_export), d = P;
      _.rendersession_source_span(o, this.__wbg_ptr, u, d);
      var e = h().getInt32(o + 0, !0), n = h().getInt32(o + 4, !0), i = h().getInt32(o + 8, !0), s = h().getInt32(o + 12, !0);
      if (s)
        throw M(i);
      let g;
      return e !== 0 && (g = E(e, n).slice(), _.__wbindgen_export4(e, n * 1, 1)), g;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (j.prototype[Symbol.dispose] = j.prototype.free);
class Pt {
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, re.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_rendersessionoptions_free(t, 0);
  }
  constructor() {
    const t = _.rendersessionoptions_new();
    return this.__wbg_ptr = t >>> 0, re.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @returns {number | undefined}
   */
  get pixel_per_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const t = _.rendersession_pixel_per_pt(this.__wbg_ptr);
    return t === 4294967297 ? void 0 : t;
  }
  /**
   * @param {number | null} [pixel_per_pt]
   */
  set pixel_per_pt(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr), F(t) || b(t), _.rendersession_set_pixel_per_pt(this.__wbg_ptr, F(t) ? 4294967297 : Math.fround(t));
  }
  /**
   * @returns {string | undefined}
   */
  get background_color() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), _.rendersessionoptions_background_color(n, this.__wbg_ptr);
      var t = h().getInt32(n + 0, !0), e = h().getInt32(n + 4, !0);
      let i;
      return t !== 0 && (i = E(t, e).slice(), _.__wbindgen_export4(t, e * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [background_color]
   */
  set background_color(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    var e = F(t) ? 0 : W(t, _.__wbindgen_export, _.__wbindgen_export2), n = P;
    _.rendersessionoptions_set_background_color(this.__wbg_ptr, e, n);
  }
  /**
   * @returns {string | undefined}
   */
  get format() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), _.rendersessionoptions_format(n, this.__wbg_ptr);
      var t = h().getInt32(n + 0, !0), e = h().getInt32(n + 4, !0);
      let i;
      return t !== 0 && (i = E(t, e).slice(), _.__wbindgen_export4(t, e * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [format]
   */
  set format(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    var e = F(t) ? 0 : W(t, _.__wbindgen_export, _.__wbindgen_export2), n = P;
    _.rendersessionoptions_set_format(this.__wbg_ptr, e, n);
  }
}
Symbol.dispose && (Pt.prototype[Symbol.dispose] = Pt.prototype.free);
class ot {
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(ot.prototype);
    return e.__wbg_ptr = t, At.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, At.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_typstrenderer_free(t, 0);
  }
  constructor() {
    const t = _.typstrenderer_new();
    return this.__wbg_ptr = t >>> 0, At.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @param {CreateSessionOptions | null} [options]
   * @returns {RenderSession}
   */
  create_session(t) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const s = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr);
      let o = 0;
      if (!F(t)) {
        if (U(t, kt), t.__wbg_ptr === 0)
          throw new Error("Attempt to use a moved value");
        o = t.__destroy_into_raw();
      }
      _.typstrenderer_create_session(s, this.__wbg_ptr, o);
      var e = h().getInt32(s + 0, !0), n = h().getInt32(s + 4, !0), i = h().getInt32(s + 8, !0);
      if (i)
        throw M(n);
      return j.__wrap(e);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} session
   */
  reset(t) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const i = _.__wbindgen_add_to_stack_pointer(-16);
      if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      _.typstrenderer_reset(i, this.__wbg_ptr, t.__wbg_ptr);
      var e = h().getInt32(i + 0, !0), n = h().getInt32(i + 4, !0);
      if (n)
        throw M(e);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} session
   * @param {string} action
   * @param {Uint8Array} data
   */
  manipulate_data(t, e, n) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      const u = W(e, _.__wbindgen_export, _.__wbindgen_export2), d = P, g = It(n, _.__wbindgen_export), w = P;
      _.typstrenderer_manipulate_data(o, this.__wbg_ptr, t.__wbg_ptr, u, d, g, w);
      var i = h().getInt32(o + 0, !0), s = h().getInt32(o + 4, !0);
      if (s)
        throw M(i);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {Uint8Array} artifact_content
   * @param {string} decoder
   * @returns {RenderSession}
   */
  session_from_artifact(t, e) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr);
      const u = It(t, _.__wbindgen_export), d = P, g = W(e, _.__wbindgen_export, _.__wbindgen_export2), w = P;
      _.typstrenderer_session_from_artifact(o, this.__wbg_ptr, u, d, g, w);
      var n = h().getInt32(o + 0, !0), i = h().getInt32(o + 4, !0), s = h().getInt32(o + 8, !0);
      if (s)
        throw M(i);
      return j.__wrap(n);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} _v
   */
  load_glyph_pack(t) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const i = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr), _.typstrenderer_load_glyph_pack(i, this.__wbg_ptr, v(t));
      var e = h().getInt32(i + 0, !0), n = h().getInt32(i + 4, !0);
      if (n)
        throw M(e);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} session
   * @param {number} rect_lo_x
   * @param {number} rect_lo_y
   * @param {number} rect_hi_x
   * @param {number} rect_hi_y
   * @returns {string}
   */
  render_svg_diff(t, e, n, i, s) {
    let o, u;
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const w = _.__wbindgen_add_to_stack_pointer(-16);
      if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      _.typstrenderer_render_svg_diff(w, this.__wbg_ptr, t.__wbg_ptr, e, n, i, s);
      var d = h().getInt32(w + 0, !0), g = h().getInt32(w + 4, !0);
      return o = d, u = g, E(d, g);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16), _.__wbindgen_export4(o, u, 1);
    }
  }
  /**
   * @param {RenderSession} session
   * @param {number | null} [parts]
   * @returns {string}
   */
  svg_data(t, e) {
    let n, i;
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const m = _.__wbindgen_add_to_stack_pointer(-16);
      if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      F(e) || b(e), _.typstrenderer_svg_data(m, this.__wbg_ptr, t.__wbg_ptr, F(e) ? 4294967297 : e >>> 0);
      var s = h().getInt32(m + 0, !0), o = h().getInt32(m + 4, !0), u = h().getInt32(m + 8, !0), d = h().getInt32(m + 12, !0), g = s, w = o;
      if (d)
        throw g = 0, w = 0, M(u);
      return n = g, i = w, E(g, w);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16), _.__wbindgen_export4(n, i, 1);
    }
  }
  /**
   * @param {RenderSession} session
   * @returns {Array<any> | undefined}
   */
  get_customs(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
      throw new Error("Attempt to use a moved value");
    const e = _.typstrenderer_get_customs(this.__wbg_ptr, t.__wbg_ptr);
    return M(e);
  }
  /**
   * @param {RenderSession} session
   * @param {HTMLElement} root
   * @returns {boolean}
   */
  render_svg(t, e) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      _.typstrenderer_render_svg(o, this.__wbg_ptr, t.__wbg_ptr, v(e));
      var n = h().getInt32(o + 0, !0), i = h().getInt32(o + 4, !0), s = h().getInt32(o + 8, !0);
      if (s)
        throw M(i);
      return n !== 0;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} _w
   * @returns {Promise<TypstWorker>}
   */
  create_worker(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const e = _.typstrenderer_create_worker(this.__wbg_ptr, v(t));
    return M(e);
  }
  /**
   * @returns {WorkerBridge}
   */
  create_worker_bridge() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const i = this.__destroy_into_raw(), s = _.__wbindgen_add_to_stack_pointer(-16);
      b(i), _.typstrenderer_create_worker_bridge(s, i);
      var t = h().getInt32(s + 0, !0), e = h().getInt32(s + 4, !0), n = h().getInt32(s + 8, !0);
      if (n)
        throw M(e);
      return _t.__wrap(t);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} ses
   * @param {any} canvas
   * @param {RenderPageImageOptions | null} [options]
   * @returns {Promise<any>}
   */
  render_page_to_canvas(t, e, n) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
      throw new Error("Attempt to use a moved value");
    let i = 0;
    if (!F(n)) {
      if (U(n, K), n.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      i = n.__destroy_into_raw();
    }
    const s = _.typstrenderer_render_page_to_canvas(this.__wbg_ptr, t.__wbg_ptr, v(e), i);
    return M(s);
  }
  /**
   * @param {RenderSession} ses
   * @param {HTMLElement} elem
   * @returns {Promise<IncrDomDocClient>}
   */
  mount_dom(t, e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    if (b(this.__wbg_ptr), U(t, j), t.__wbg_ptr === 0)
      throw new Error("Attempt to use a moved value");
    const n = _.typstrenderer_mount_dom(this.__wbg_ptr, t.__wbg_ptr, v(e));
    return M(n);
  }
}
Symbol.dispose && (ot.prototype[Symbol.dispose] = ot.prototype.free);
class Tt {
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ne.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_typstrendererbuilder_free(t, 0);
  }
  constructor() {
    try {
      const i = _.__wbindgen_add_to_stack_pointer(-16);
      _.typstrendererbuilder_new(i);
      var t = h().getInt32(i + 0, !0), e = h().getInt32(i + 4, !0), n = h().getInt32(i + 8, !0);
      if (n)
        throw M(e);
      return this.__wbg_ptr = t >>> 0, ne.register(this, this.__wbg_ptr, this), this;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @returns {Promise<TypstRenderer>}
   */
  build() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    const t = this.__destroy_into_raw();
    b(t);
    const e = _.typstrendererbuilder_build(t);
    return M(e);
  }
  /**
   * @param {any} _pack
   * @returns {Promise<void>}
   */
  add_glyph_pack(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const e = _.typstrendererbuilder_add_glyph_pack(this.__wbg_ptr, v(t));
    return M(e);
  }
  /**
   * @param {Uint8Array} _font_buffer
   * @returns {Promise<void>}
   */
  add_raw_font(t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const e = _.typstrendererbuilder_add_raw_font(this.__wbg_ptr, v(t));
    return M(e);
  }
  /**
   * @param {any} _font
   * @param {any} _blob
   * @returns {Promise<void>}
   */
  add_lazy_font(t, e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const n = _.typstrendererbuilder_add_lazy_font(this.__wbg_ptr, v(t), v(e));
    return M(n);
  }
}
Symbol.dispose && (Tt.prototype[Symbol.dispose] = Tt.prototype.free);
class Ot {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, vr.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_typstworker_free(t, 0);
  }
  /**
   * @param {string} _action
   * @param {Uint8Array} _data
   * @returns {Promise<any>}
   */
  manipulate_data(t, e) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr);
      const u = W(t, _.__wbindgen_export, _.__wbindgen_export2), d = P;
      _.typstworker_manipulate_data(o, this.__wbg_ptr, u, d, v(e));
      var n = h().getInt32(o + 0, !0), i = h().getInt32(o + 4, !0), s = h().getInt32(o + 8, !0);
      if (s)
        throw M(i);
      return M(n);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @returns {Promise<any>}
   */
  get_pages_info() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    b(this.__wbg_ptr);
    const t = _.typstworker_get_pages_info(this.__wbg_ptr);
    return M(t);
  }
  /**
   * @param {Uint8Array} _actions
   * @param {HTMLCanvasElement[]} _canvas_list
   * @param {RenderPageImageOptions[]} _data
   * @returns {Promise<any>}
   */
  render_canvas(t, e, n) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const u = _.__wbindgen_add_to_stack_pointer(-16);
      b(this.__wbg_ptr);
      const d = It(t, _.__wbindgen_export), g = P, w = Gt(e, _.__wbindgen_export), m = P, C = Gt(n, _.__wbindgen_export), R = P;
      _.typstworker_render_canvas(u, this.__wbg_ptr, d, g, w, m, C, R);
      var i = h().getInt32(u + 0, !0), s = h().getInt32(u + 4, !0), o = h().getInt32(u + 8, !0);
      if (o)
        throw M(s);
      return M(i);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (Ot.prototype[Symbol.dispose] = Ot.prototype.free);
class _t {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(_t.prototype);
    return e.__wbg_ptr = t, ie.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ie.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    _.__wbg_workerbridge_free(t, 0);
  }
}
Symbol.dispose && (_t.prototype[Symbol.dispose] = _t.prototype.free);
function Sr() {
  const r = _.renderer_build_info();
  return M(r);
}
const Ir = /* @__PURE__ */ new Set(["basic", "cors", "default"]);
async function kr(r, t) {
  if (typeof Response == "function" && r instanceof Response) {
    if (typeof WebAssembly.instantiateStreaming == "function")
      try {
        return await WebAssembly.instantiateStreaming(r, t);
      } catch (n) {
        if (r.ok && Ir.has(r.type) && r.headers.get("Content-Type") !== "application/wasm")
          console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve Wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", n);
        else
          throw n;
      }
    const e = await r.arrayBuffer();
    return await WebAssembly.instantiate(e, t);
  } else {
    const e = await WebAssembly.instantiate(r, t);
    return e instanceof WebAssembly.Instance ? { instance: e, module: r } : e;
  }
}
function Se() {
  const r = {};
  return r.wbg = {}, r.wbg.__wbg___wbindgen_debug_string_adfb662ae34724b6 = function(t, e) {
    const n = Mt(c(e)), i = W(n, _.__wbindgen_export, _.__wbindgen_export2), s = P;
    h().setInt32(t + 4, s, !0), h().setInt32(t + 0, i, !0);
  }, r.wbg.__wbg___wbindgen_is_function_8d400b8b1af978cd = function(t) {
    const e = typeof c(t) == "function";
    return D(e), e;
  }, r.wbg.__wbg___wbindgen_is_undefined_f6b95eab589e0269 = function(t) {
    const e = c(t) === void 0;
    return D(e), e;
  }, r.wbg.__wbg___wbindgen_jsval_eq_b6101cc9cef1fe36 = function(t, e) {
    const n = c(t) === c(e);
    return D(n), n;
  }, r.wbg.__wbg___wbindgen_string_get_a2a31e16edf96e42 = function(t, e) {
    const n = c(e), i = typeof n == "string" ? n : void 0;
    var s = F(i) ? 0 : W(i, _.__wbindgen_export, _.__wbindgen_export2), o = P;
    h().setInt32(t + 4, o, !0), h().setInt32(t + 0, s, !0);
  }, r.wbg.__wbg___wbindgen_throw_dd24417ed36fc46e = function(t, e) {
    throw new Error(E(t, e));
  }, r.wbg.__wbg__wbg_cb_unref_87dfb5aaa0cbcea7 = function() {
    return f(function(t) {
      c(t)._wbg_cb_unref();
    }, arguments);
  }, r.wbg.__wbg_appendChild_7465eba84213c75f = function() {
    return x(function(t, e) {
      const n = c(t).appendChild(c(e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_call_3020136f7a2d6e44 = function() {
    return x(function(t, e, n) {
      const i = c(t).call(c(e), c(n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_call_78f94eb02ec7f9b2 = function() {
    return x(function(t, e, n, i, s) {
      const o = c(t).call(c(e), c(n), c(i), c(s));
      return v(o);
    }, arguments);
  }, r.wbg.__wbg_call_abb4ff46ce38be40 = function() {
    return x(function(t, e) {
      const n = c(t).call(c(e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_call_c8baa5c5e72d274e = function() {
    return x(function(t, e, n, i) {
      const s = c(t).call(c(e), c(n), c(i));
      return v(s);
    }, arguments);
  }, r.wbg.__wbg_clearRect_c2909c2f38887f82 = function() {
    return f(function(t, e, n, i, s) {
      c(t).clearRect(e, n, i, s);
    }, arguments);
  }, r.wbg.__wbg_clientWidth_dbc9540f4ebdce2a = function() {
    return f(function(t) {
      const e = c(t).clientWidth;
      return b(e), e;
    }, arguments);
  }, r.wbg.__wbg_clip_35de02bb25d222d9 = function() {
    return f(function(t, e) {
      c(t).clip(c(e));
    }, arguments);
  }, r.wbg.__wbg_clip_afc9a92171eee96e = function() {
    return f(function(t, e) {
      c(t).clip(c(e));
    }, arguments);
  }, r.wbg.__wbg_cloneNode_c9c45b24b171a776 = function() {
    return x(function(t) {
      const e = c(t).cloneNode();
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_content_ad90fa08b8c037c5 = function() {
    return f(function(t) {
      const e = c(t).content;
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_createElement_da4ed2b219560fc6 = function() {
    return x(function(t, e, n) {
      const i = c(t).createElement(E(e, n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_createImageBitmap_15b94d7a7277d0be = function() {
    return x(function(t, e) {
      const n = c(t).createImageBitmap(c(e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_createImageBitmap_4772db6cb0dff013 = function() {
    return x(function(t, e) {
      const n = c(t).createImageBitmap(c(e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_createObjectURL_7d9f7f8f41373850 = function() {
    return x(function(t, e) {
      const n = URL.createObjectURL(c(e)), i = W(n, _.__wbindgen_export, _.__wbindgen_export2), s = P;
      h().setInt32(t + 4, s, !0), h().setInt32(t + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_document_5b745e82ba551ca5 = function() {
    return f(function(t) {
      const e = c(t).document;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_drawImage_2859dfb229439d8c = function() {
    return x(function(t, e, n, i, s, o) {
      c(t).drawImage(c(e), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_drawImage_6949f7679b2a9758 = function() {
    return x(function(t, e, n, i, s, o) {
      c(t).drawImage(c(e), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_drawImage_c2aacfe3133aedb7 = function() {
    return x(function(t, e, n, i) {
      c(t).drawImage(c(e), n, i);
    }, arguments);
  }, r.wbg.__wbg_drawImage_c40aff4a38206f4a = function() {
    return x(function(t, e, n, i, s, o) {
      c(t).drawImage(c(e), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_drawImage_c4f2c6d2bcfa9463 = function() {
    return x(function(t, e, n, i) {
      c(t).drawImage(c(e), n, i);
    }, arguments);
  }, r.wbg.__wbg_drawImage_de5946f9cc23c2b9 = function() {
    return x(function(t, e, n, i) {
      c(t).drawImage(c(e), n, i);
    }, arguments);
  }, r.wbg.__wbg_drawImage_e2012c2a2d434ee2 = function() {
    return x(function(t, e, n, i, s, o) {
      c(t).drawImage(c(e), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_error_7534b8e9a36f1ab4 = function() {
    return f(function(t, e) {
      let n, i;
      try {
        n = t, i = e, console.error(E(t, e));
      } finally {
        _.__wbindgen_export4(n, i, 1);
      }
    }, arguments);
  }, r.wbg.__wbg_fillRect_84131220403e26a4 = function() {
    return f(function(t, e, n, i, s) {
      c(t).fillRect(e, n, i, s);
    }, arguments);
  }, r.wbg.__wbg_fillRect_e0efe328740308e7 = function() {
    return f(function(t, e, n, i, s) {
      c(t).fillRect(e, n, i, s);
    }, arguments);
  }, r.wbg.__wbg_fill_2f2cb4f3f75aadd4 = function() {
    return f(function(t, e, n) {
      c(t).fill(c(e), Xt[n]);
    }, arguments);
  }, r.wbg.__wbg_fill_38a7d7455451c001 = function() {
    return f(function(t, e, n) {
      c(t).fill(c(e), Xt[n]);
    }, arguments);
  }, r.wbg.__wbg_fill_3d82e52894462c95 = function() {
    return f(function(t, e) {
      c(t).fill(c(e));
    }, arguments);
  }, r.wbg.__wbg_fill_f1043be5324533f3 = function() {
    return f(function(t, e) {
      c(t).fill(c(e));
    }, arguments);
  }, r.wbg.__wbg_firstElementChild_54d515b19cee5ae7 = function() {
    return f(function(t) {
      const e = c(t).firstElementChild;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_firstElementChild_e207b33aaa4a86df = function() {
    return f(function(t) {
      const e = c(t).firstElementChild;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_getAttribute_80900eec94cb3636 = function() {
    return f(function(t, e, n, i) {
      const s = c(e).getAttribute(E(n, i));
      var o = F(s) ? 0 : W(s, _.__wbindgen_export, _.__wbindgen_export2), u = P;
      h().setInt32(t + 4, u, !0), h().setInt32(t + 0, o, !0);
    }, arguments);
  }, r.wbg.__wbg_getContext_01f42b234e833f0a = function() {
    return x(function(t, e, n) {
      const i = c(t).getContext(E(e, n));
      return F(i) ? 0 : v(i);
    }, arguments);
  }, r.wbg.__wbg_getContext_2f210d0a58d43d95 = function() {
    return x(function(t, e, n) {
      const i = c(t).getContext(E(e, n));
      return F(i) ? 0 : v(i);
    }, arguments);
  }, r.wbg.__wbg_get_af9dab7e9603ea93 = function() {
    return x(function(t, e) {
      const n = Reflect.get(c(t), c(e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_globalCompositeOperation_03e6a5201c8ee369 = function() {
    return x(function(t, e) {
      const n = c(e).globalCompositeOperation, i = W(n, _.__wbindgen_export, _.__wbindgen_export2), s = P;
      h().setInt32(t + 4, s, !0), h().setInt32(t + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_globalCompositeOperation_d52c6e6666e4be00 = function() {
    return x(function(t, e) {
      const n = c(e).globalCompositeOperation, i = W(n, _.__wbindgen_export, _.__wbindgen_export2), s = P;
      h().setInt32(t + 4, s, !0), h().setInt32(t + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_height_3a0b31e52a5b7f17 = function() {
    return f(function(t) {
      const e = c(t).height;
      return b(e), e;
    }, arguments);
  }, r.wbg.__wbg_incrdomdocclient_new = function() {
    return f(function(t) {
      const e = it.__wrap(t);
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_instanceof_CanvasRenderingContext2d_d070139aaac1459f = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof CanvasRenderingContext2D;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_Element_6f7ba982258cfc0f = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof Element;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlCanvasElement_c4251b1b6a15edcc = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof HTMLCanvasElement;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlDivElement_a55b9d453d2d8037 = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof HTMLDivElement;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlElement_20a3acb594113d73 = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof HTMLElement;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlTemplateElement_702a22ea7e74d089 = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof HTMLTemplateElement;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_ImageBitmap_e44697c70dffc0fe = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof ImageBitmap;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_OffscreenCanvasRenderingContext2d_52996da29a2bf2f5 = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof OffscreenCanvasRenderingContext2D;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_OffscreenCanvas_8fb961f5ea27473f = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof OffscreenCanvas;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_Promise_eca6c43a2610558d = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof Promise;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_SvgGraphicsElement_16c442d1a36270b3 = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof SVGGraphicsElement;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_SvgsvgElement_27a38efc9750b6fa = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof SVGSVGElement;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_Window_b5cf7783caa68180 = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof Window;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_WorkerGlobalScope_9a3411db21c65a54 = function() {
    return f(function(t) {
      let e;
      try {
        e = c(t) instanceof WorkerGlobalScope;
      } catch {
        e = !1;
      }
      const n = e;
      return D(n), n;
    }, arguments);
  }, r.wbg.__wbg_lastElementChild_6b90fe10adf639a2 = function() {
    return f(function(t) {
      const e = c(t).lastElementChild;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_length_22ac23eaec9d8053 = function() {
    return f(function(t) {
      const e = c(t).length;
      return b(e), e;
    }, arguments);
  }, r.wbg.__wbg_log_1d990106d99dacb7 = function() {
    return f(function(t) {
      console.log(c(t));
    }, arguments);
  }, r.wbg.__wbg_log_fd6486c6d5396ce5 = function() {
    return f(function(t, e) {
      console.log(c(t), c(e));
    }, arguments);
  }, r.wbg.__wbg_measureText_a465ecc06491802d = function() {
    return x(function(t, e, n) {
      const i = c(t).measureText(E(e, n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_new_1ba21ce319a06297 = function() {
    return f(function() {
      const t = new Object();
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_new_25f239778d6112b9 = function() {
    return f(function() {
      const t = new Array();
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_new_2884ddbcda091bc1 = function() {
    return x(function() {
      const t = new Image();
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_new_8a6f238a6ece86ea = function() {
    return f(function() {
      const t = new Error();
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_new_9468dd6e5df427f6 = function() {
    return x(function(t, e) {
      const n = new OffscreenCanvas(t >>> 0, e >>> 0);
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_new_df1173567d5ff028 = function() {
    return f(function(t, e) {
      const n = new Error(E(t, e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_new_ff12d2b041fb48f1 = function() {
    return f(function(t, e) {
      try {
        var n = { a: t, b: e }, i = (o, u) => {
          const d = n.a;
          n.a = 0;
          try {
            return yr(d, n.b, o, u);
          } finally {
            n.a = d;
          }
        };
        const s = new Promise(i);
        return v(s);
      } finally {
        n.a = n.b = 0;
      }
    }, arguments);
  }, r.wbg.__wbg_new_from_slice_f9c22b9153b26992 = function() {
    return f(function(t, e) {
      const n = new Uint8Array(Jt(t, e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_new_no_args_cb138f77cf6151ee = function() {
    return f(function(t, e) {
      const n = new Function(E(t, e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_new_with_length_aa5eaf41d35235e5 = function() {
    return f(function(t) {
      const e = new Uint8Array(t >>> 0);
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_new_with_path_string_3aea637e78135c46 = function() {
    return x(function(t, e) {
      const n = new Path2D(E(t, e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_new_with_u8_array_sequence_and_options_d4def9ec0588c7ec = function() {
    return x(function(t, e) {
      const n = new Blob(c(t), c(e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_nextElementSibling_c745ee48314963c0 = function() {
    return f(function(t) {
      const e = c(t).nextElementSibling;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_push_7d9be8f38fc13975 = function() {
    return f(function(t, e) {
      const n = c(t).push(c(e));
      return b(n), n;
    }, arguments);
  }, r.wbg.__wbg_putImageData_c280ca107c4b7828 = function() {
    return x(function(t, e, n, i) {
      c(t).putImageData(c(e), n, i);
    }, arguments);
  }, r.wbg.__wbg_putImageData_e96eabf0e9610ed3 = function() {
    return x(function(t, e, n, i) {
      c(t).putImageData(c(e), n, i);
    }, arguments);
  }, r.wbg.__wbg_queueMicrotask_9b549dfce8865860 = function() {
    return f(function(t) {
      const e = c(t).queueMicrotask;
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_queueMicrotask_fca69f5bfad613a5 = function() {
    return f(function(t) {
      queueMicrotask(c(t));
    }, arguments);
  }, r.wbg.__wbg_removeProperty_c2e16faee2834bef = function() {
    return x(function(t, e, n, i) {
      const s = c(e).removeProperty(E(n, i)), o = W(s, _.__wbindgen_export, _.__wbindgen_export2), u = P;
      h().setInt32(t + 4, u, !0), h().setInt32(t + 0, o, !0);
    }, arguments);
  }, r.wbg.__wbg_remove_32f69ffabcbc4072 = function() {
    return f(function(t) {
      c(t).remove();
    }, arguments);
  }, r.wbg.__wbg_renderpageimageoptions_unwrap = function() {
    return f(function(t) {
      const e = K.__unwrap(c(t));
      return b(e), e;
    }, arguments);
  }, r.wbg.__wbg_replaceWith_923cfb1b2f584408 = function() {
    return x(function(t, e) {
      c(t).replaceWith(c(e));
    }, arguments);
  }, r.wbg.__wbg_resolve_fd5bfbaa4ce36e1e = function() {
    return f(function(t) {
      const e = Promise.resolve(c(t));
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_restore_352c39c9bbeedc91 = function() {
    return f(function(t) {
      c(t).restore();
    }, arguments);
  }, r.wbg.__wbg_restore_6486cb1a7aa3af7b = function() {
    return f(function(t) {
      c(t).restore();
    }, arguments);
  }, r.wbg.__wbg_revokeObjectURL_88db3468842ff09e = function() {
    return x(function(t, e) {
      URL.revokeObjectURL(E(t, e));
    }, arguments);
  }, r.wbg.__wbg_save_131c8dc648f702b6 = function() {
    return f(function(t) {
      c(t).save();
    }, arguments);
  }, r.wbg.__wbg_save_b8767cfd2ee7f600 = function() {
    return f(function(t) {
      c(t).save();
    }, arguments);
  }, r.wbg.__wbg_setAttribute_34747dd193f45828 = function() {
    return x(function(t, e, n, i, s) {
      c(t).setAttribute(E(e, n), E(i, s));
    }, arguments);
  }, r.wbg.__wbg_setLineDash_7e3aad159aa2710c = function() {
    return x(function(t, e) {
      c(t).setLineDash(c(e));
    }, arguments);
  }, r.wbg.__wbg_setLineDash_cfbc3a6bfddeee62 = function() {
    return x(function(t, e) {
      c(t).setLineDash(c(e));
    }, arguments);
  }, r.wbg.__wbg_setProperty_f27b2c05323daf8a = function() {
    return x(function(t, e, n, i, s) {
      c(t).setProperty(E(e, n), E(i, s));
    }, arguments);
  }, r.wbg.__wbg_setTransform_9de588d9f080f5bc = function() {
    return x(function(t, e, n, i, s, o, u) {
      c(t).setTransform(e, n, i, s, o, u);
    }, arguments);
  }, r.wbg.__wbg_setTransform_c55eca1193bf5ea1 = function() {
    return x(function(t, e, n, i, s, o, u) {
      c(t).setTransform(e, n, i, s, o, u);
    }, arguments);
  }, r.wbg.__wbg_set_169e13b608078b7b = function() {
    return f(function(t, e, n) {
      c(t).set(Jt(e, n));
    }, arguments);
  }, r.wbg.__wbg_set_781438a03c0c3c81 = function() {
    return x(function(t, e, n) {
      const i = Reflect.set(c(t), c(e), c(n));
      return D(i), i;
    }, arguments);
  }, r.wbg.__wbg_set_7df433eea03a5c14 = function() {
    return f(function(t, e, n) {
      c(t)[e >>> 0] = M(n);
    }, arguments);
  }, r.wbg.__wbg_set_fillStyle_b26e462a87b14315 = function() {
    return f(function(t, e, n) {
      c(t).fillStyle = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_fillStyle_c9a0550307cd4671 = function() {
    return f(function(t, e, n) {
      c(t).fillStyle = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_font_6d67b15564a1e344 = function() {
    return f(function(t, e, n) {
      c(t).font = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_globalCompositeOperation_418a0b97ba91a947 = function() {
    return x(function(t, e, n) {
      c(t).globalCompositeOperation = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_globalCompositeOperation_7e16c72877ba5f86 = function() {
    return x(function(t, e, n) {
      c(t).globalCompositeOperation = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_height_6f8f8ef4cb40e496 = function() {
    return f(function(t, e) {
      c(t).height = e >>> 0;
    }, arguments);
  }, r.wbg.__wbg_set_innerHTML_f1d03f780518a596 = function() {
    return f(function(t, e, n) {
      c(t).innerHTML = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_lineCap_74b37c7ff968f854 = function() {
    return f(function(t, e, n) {
      c(t).lineCap = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_lineCap_791e7648138cc371 = function() {
    return f(function(t, e, n) {
      c(t).lineCap = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_lineDashOffset_85f21a63ddf34d2c = function() {
    return f(function(t, e) {
      c(t).lineDashOffset = e;
    }, arguments);
  }, r.wbg.__wbg_set_lineDashOffset_f032b4e2a273a2f5 = function() {
    return f(function(t, e) {
      c(t).lineDashOffset = e;
    }, arguments);
  }, r.wbg.__wbg_set_lineJoin_196c6ac02fd494c3 = function() {
    return f(function(t, e, n) {
      c(t).lineJoin = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_lineJoin_a136f09dd803e0cc = function() {
    return f(function(t, e, n) {
      c(t).lineJoin = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_lineWidth_ed39767bf75d602d = function() {
    return f(function(t, e) {
      c(t).lineWidth = e;
    }, arguments);
  }, r.wbg.__wbg_set_lineWidth_feda4b79a15c660b = function() {
    return f(function(t, e) {
      c(t).lineWidth = e;
    }, arguments);
  }, r.wbg.__wbg_set_miterLimit_b12c03757937d228 = function() {
    return f(function(t, e) {
      c(t).miterLimit = e;
    }, arguments);
  }, r.wbg.__wbg_set_miterLimit_d891f47e244a915c = function() {
    return f(function(t, e) {
      c(t).miterLimit = e;
    }, arguments);
  }, r.wbg.__wbg_set_onerror_e7e40c62a55a0770 = function() {
    return f(function(t, e) {
      c(t).onerror = c(e);
    }, arguments);
  }, r.wbg.__wbg_set_onload_5e2862e3453854de = function() {
    return f(function(t, e) {
      c(t).onload = c(e);
    }, arguments);
  }, r.wbg.__wbg_set_src_84f27c5105946dce = function() {
    return f(function(t, e, n) {
      c(t).src = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_strokeStyle_25d5381be570e780 = function() {
    return f(function(t, e, n) {
      c(t).strokeStyle = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_strokeStyle_697a576d2d3fbeaa = function() {
    return f(function(t, e, n) {
      c(t).strokeStyle = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_type_7ce650670a34c68f = function() {
    return f(function(t, e, n) {
      c(t).type = E(e, n);
    }, arguments);
  }, r.wbg.__wbg_set_width_7ff7a22c6e9f423e = function() {
    return f(function(t, e) {
      c(t).width = e >>> 0;
    }, arguments);
  }, r.wbg.__wbg_stack_0ed75d68575b0f3c = function() {
    return f(function(t, e) {
      const n = c(e).stack, i = W(n, _.__wbindgen_export, _.__wbindgen_export2), s = P;
      h().setInt32(t + 4, s, !0), h().setInt32(t + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_GLOBAL_769e6b65d6557335 = function() {
    return f(function() {
      const t = typeof global > "u" ? null : global;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_GLOBAL_THIS_60cf02db4de8e1c1 = function() {
    return f(function() {
      const t = typeof globalThis > "u" ? null : globalThis;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_SELF_08f5a74c69739274 = function() {
    return f(function() {
      const t = typeof self > "u" ? null : self;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_WINDOW_a8924b26aa92d024 = function() {
    return f(function() {
      const t = typeof window > "u" ? null : window;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_stringify_655a6390e1f5eb6b = function() {
    return x(function(t) {
      const e = JSON.stringify(c(t));
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_stroke_41b0736a6a0b61aa = function() {
    return f(function(t, e) {
      c(t).stroke(c(e));
    }, arguments);
  }, r.wbg.__wbg_stroke_7cfe6d60a6c48303 = function() {
    return f(function(t, e) {
      c(t).stroke(c(e));
    }, arguments);
  }, r.wbg.__wbg_style_521a717da50e53c6 = function() {
    return f(function(t) {
      const e = c(t).style;
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_then_429f7caf1026411d = function() {
    return f(function(t, e, n) {
      const i = c(t).then(c(e), c(n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_then_4f95312d68691235 = function() {
    return f(function(t, e) {
      const n = c(t).then(c(e));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_transferToImageBitmap_7048556ff33da7ed = function() {
    return x(function(t) {
      const e = c(t).transferToImageBitmap();
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_typstrenderer_new = function() {
    return f(function(t) {
      const e = ot.__wrap(t);
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_warn_6e567d0d926ff881 = function() {
    return f(function(t) {
      console.warn(c(t));
    }, arguments);
  }, r.wbg.__wbg_width_c3f96f2ca7ab208d = function() {
    return f(function(t) {
      return c(t).width;
    }, arguments);
  }, r.wbg.__wbg_width_f13d2e86324fc226 = function() {
    return f(function(t) {
      const e = c(t).width;
      return b(e), e;
    }, arguments);
  }, r.wbg.__wbindgen_cast_2241b6af4c4b2941 = function() {
    return f(function(t, e) {
      const n = E(t, e);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_cast_34ce245e5be1ba8b = function() {
    return f(function(t, e) {
      const n = Vt(t, e, _.__wasm_bindgen_func_elem_1022, mr);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_cast_b2e692da0ea8478a = function() {
    return f(function(t, e) {
      const n = fr(t, e, _.__wasm_bindgen_func_elem_1077, pr);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_cast_d3d3bd81840cfbdd = function() {
    return f(function(t, e) {
      const n = Vt(t, e, _.__wasm_bindgen_func_elem_1022, hr);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_cast_d6cd19b81560fd6e = function() {
    return f(function(t) {
      return v(t);
    }, arguments);
  }, r.wbg.__wbindgen_object_clone_ref = function(t) {
    const e = c(t);
    return v(e);
  }, r.wbg.__wbindgen_object_drop_ref = function(t) {
    M(t);
  }, r;
}
function Ie(r, t) {
  return _ = r.exports, ke.__wbindgen_wasm_module = t, V = null, dt = null, gt = null, _;
}
function xr(r) {
  if (_ !== void 0) return _;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module: r } = r : console.warn("using deprecated parameters for `initSync()`; pass a single object instead"));
  const t = Se();
  r instanceof WebAssembly.Module || (r = new WebAssembly.Module(r));
  const e = new WebAssembly.Instance(r, t);
  return Ie(e, r);
}
async function ke(r) {
  if (_ !== void 0) return _;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module_or_path: r } = r : console.warn("using deprecated parameters for the initialization function; pass a single object instead")), typeof r > "u" && (r = xe("typst_ts_renderer_bg.wasm", import.meta.url));
  const t = Se();
  (typeof r == "string" || typeof Request == "function" && r instanceof Request || typeof URL == "function" && r instanceof URL) && (r = fetch(r));
  const { instance: e, module: n } = await kr(await r, t);
  return Ie(e, n);
}
let xe = async function(r, t) {
  throw new Error("Cannot import wasm module without importer: " + r + " " + t);
};
function Ee(r) {
  xe = r;
}
let Er = async function(r, t) {
  const e = new Function("m", "return import(m)"), { readFileSync: n } = await e("fs"), i = new URL(r, t);
  return await n(i).buffer;
};
const Rr = typeof process < "u" && process.versions != null && process.versions.node != null;
Rr && Ee(Er);
const Re = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  CreateSessionOptions: kt,
  IncrDomDocClient: it,
  PageInfo: X,
  PagesInfo: st,
  RenderPageImageOptions: K,
  RenderSession: j,
  RenderSessionOptions: Pt,
  TypstRenderer: ot,
  TypstRendererBuilder: Tt,
  TypstWorker: Ot,
  WorkerBridge: _t,
  default: ke,
  initSync: xr,
  renderer_build_info: Sr,
  setImportWasmModule: Ee
}, Symbol.toStringTag, { value: "Module" }));
let a;
function S(r) {
  bt === H.length && H.push(H.length + 1);
  const t = bt;
  return bt = H[t], H[t] = r, t;
}
function Lt(r, t) {
  if (!(r instanceof t))
    throw new Error(`expected instance of ${t.name}`);
}
const se = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => r.dtor(r.a, r.b));
function $t(r) {
  const t = typeof r;
  if (t == "number" || t == "boolean" || r == null)
    return `${r}`;
  if (t == "string")
    return `"${r}"`;
  if (t == "symbol") {
    const i = r.description;
    return i == null ? "Symbol" : `Symbol(${i})`;
  }
  if (t == "function") {
    const i = r.name;
    return typeof i == "string" && i.length > 0 ? `Function(${i})` : "Function";
  }
  if (Array.isArray(r)) {
    const i = r.length;
    let s = "[";
    i > 0 && (s += $t(r[0]));
    for (let o = 1; o < i; o++)
      s += ", " + $t(r[o]);
    return s += "]", s;
  }
  const e = /\[object ([^\]]+)\]/.exec(toString.call(r));
  let n;
  if (e && e.length > 1)
    n = e[1];
  else
    return toString.call(r);
  if (n == "Object")
    try {
      return "Object(" + JSON.stringify(r) + ")";
    } catch {
      return "Object";
    }
  return r instanceof Error ? `${r.name}: ${r.message}
${r.stack}` : n;
}
function Ar(r) {
  r < 132 || (H[r] = bt, bt = r);
}
function Cr(r, t) {
  r = r >>> 0;
  const e = l(), n = [];
  for (let i = r; i < r + 4 * t; i += 4)
    n.push(I(e.getUint32(i, !0)));
  return n;
}
function Fr(r, t) {
  return r = r >>> 0, Mr().subarray(r / 4, r / 4 + t);
}
function Dt(r, t) {
  return r = r >>> 0, nt().subarray(r / 1, r / 1 + t);
}
let G = null;
function l() {
  return (G === null || G.buffer.detached === !0 || G.buffer.detached === void 0 && G.buffer !== a.memory.buffer) && (G = new DataView(a.memory.buffer)), G;
}
function q(r, t) {
  return r = r >>> 0, Or(r, t);
}
let ft = null;
function Mr() {
  return (ft === null || ft.byteLength === 0) && (ft = new Uint32Array(a.memory.buffer)), ft;
}
let wt = null;
function nt() {
  return (wt === null || wt.byteLength === 0) && (wt = new Uint8Array(a.memory.buffer)), wt;
}
function y(r) {
  return H[r];
}
function Y(r, t) {
  try {
    return r.apply(this, t);
  } catch (e) {
    a.__wbindgen_export3(S(e));
  }
}
let H = new Array(128).fill(void 0);
H.push(void 0, null, !0, !1);
let bt = H.length;
function O(r) {
  return r == null;
}
function Pr(r, t, e, n) {
  const i = { a: r, b: t, cnt: 1, dtor: e }, s = (...o) => {
    i.cnt++;
    const u = i.a;
    i.a = 0;
    try {
      return n(u, i.b, ...o);
    } finally {
      i.a = u, s._wbg_cb_unref();
    }
  };
  return s._wbg_cb_unref = () => {
    --i.cnt === 0 && (i.dtor(i.a, i.b), i.a = 0, se.unregister(i));
  }, se.register(s, i, i), s;
}
function Ae(r, t) {
  const e = t(r.length * 1, 1) >>> 0;
  return nt().set(r, e / 1), A = r.length, e;
}
function ht(r, t) {
  const e = t(r.length * 4, 4) >>> 0, n = l();
  for (let i = 0; i < r.length; i++)
    n.setUint32(e + 4 * i, S(r[i]), !0);
  return A = r.length, e;
}
function T(r, t, e) {
  if (e === void 0) {
    const u = pt.encode(r), d = t(u.length, 1) >>> 0;
    return nt().subarray(d, d + u.length).set(u), A = u.length, d;
  }
  let n = r.length, i = t(n, 1) >>> 0;
  const s = nt();
  let o = 0;
  for (; o < n; o++) {
    const u = r.charCodeAt(o);
    if (u > 127) break;
    s[i + o] = u;
  }
  if (o !== n) {
    o !== 0 && (r = r.slice(o)), i = e(i, n, n = o + r.length * 3, 1) >>> 0;
    const u = nt().subarray(i + o, i + n), d = pt.encodeInto(r, u);
    o += d.written, i = e(i, n, o, 1) >>> 0;
  }
  return A = o, i;
}
function I(r) {
  const t = y(r);
  return Ar(r), t;
}
let yt = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 });
yt.decode();
const Tr = 2146435072;
let Ct = 0;
function Or(r, t) {
  return Ct += t, Ct >= Tr && (yt = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 }), yt.decode(), Ct = t), yt.decode(nt().subarray(r, r + t));
}
const pt = new TextEncoder();
"encodeInto" in pt || (pt.encodeInto = function(r, t) {
  const e = pt.encode(r);
  return t.set(e), {
    read: r.length,
    written: e.length
  };
});
let A = 0;
function Lr(r, t, e) {
  a.__wasm_bindgen_func_elem_944(r, t, S(e));
}
function $r(r, t, e, n) {
  a.__wasm_bindgen_func_elem_37350(r, t, S(e), S(n));
}
const oe = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => a.__wbg_incrserver_free(r >>> 0, 1)), Ft = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => a.__wbg_proxycontext_free(r >>> 0, 1)), _e = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => a.__wbg_typstcompileworld_free(r >>> 0, 1)), ae = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => a.__wbg_typstcompiler_free(r >>> 0, 1)), ce = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => a.__wbg_typstcompilerbuilder_free(r >>> 0, 1)), ue = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => a.__wbg_typstfontresolver_free(r >>> 0, 1)), de = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => a.__wbg_typstfontresolverbuilder_free(r >>> 0, 1));
class J {
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(J.prototype);
    return e.__wbg_ptr = t, oe.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, oe.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    a.__wbg_incrserver_free(t, 0);
  }
  /**
   * @param {boolean} attach
   */
  set_attach_debug_info(t) {
    a.incrserver_set_attach_debug_info(this.__wbg_ptr, t);
  }
  /**
   * @returns {Uint8Array | undefined}
   */
  current() {
    try {
      const n = a.__wbindgen_add_to_stack_pointer(-16);
      a.incrserver_current(n, this.__wbg_ptr);
      var t = l().getInt32(n + 0, !0), e = l().getInt32(n + 4, !0);
      let i;
      return t !== 0 && (i = Dt(t, e).slice(), a.__wbindgen_export4(t, e * 1, 1)), i;
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  reset() {
    a.incrserver_reset(this.__wbg_ptr);
  }
}
Symbol.dispose && (J.prototype[Symbol.dispose] = J.prototype.free);
class at {
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(at.prototype);
    return e.__wbg_ptr = t, Ft.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, Ft.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    a.__wbg_proxycontext_free(t, 0);
  }
  /**
   * Creates a new `ProxyContext` instance.
   * @param {any} context
   */
  constructor(t) {
    const e = a.proxycontext_new(S(t));
    return this.__wbg_ptr = e >>> 0, Ft.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * Returns the JavaScript this.
   * @returns {any}
   */
  get context() {
    const t = a.proxycontext_context(this.__wbg_ptr);
    return I(t);
  }
  /**
   * A convenience function to untar a tarball and call a callback for each
   * entry.
   * @param {Uint8Array} data
   * @param {Function} cb
   */
  untar(t, e) {
    try {
      const s = a.__wbindgen_add_to_stack_pointer(-16), o = Ae(t, a.__wbindgen_export), u = A;
      a.proxycontext_untar(s, this.__wbg_ptr, o, u, S(e));
      var n = l().getInt32(s + 0, !0), i = l().getInt32(s + 4, !0);
      if (i)
        throw I(n);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (at.prototype[Symbol.dispose] = at.prototype.free);
class ct {
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(ct.prototype);
    return e.__wbg_ptr = t, _e.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, _e.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    a.__wbg_typstcompileworld_free(t, 0);
  }
  /**
   * @param {number} kind
   * @param {number} diagnostics_format
   * @returns {any}
   */
  compile(t, e) {
    try {
      const o = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompileworld_compile(o, this.__wbg_ptr, t, e);
      var n = l().getInt32(o + 0, !0), i = l().getInt32(o + 4, !0), s = l().getInt32(o + 8, !0);
      if (s)
        throw I(i);
      return I(n);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} kind
   * @returns {string | undefined}
   */
  title(t) {
    try {
      const o = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompileworld_title(o, this.__wbg_ptr, t);
      var e = l().getInt32(o + 0, !0), n = l().getInt32(o + 4, !0), i = l().getInt32(o + 8, !0), s = l().getInt32(o + 12, !0);
      if (s)
        throw I(i);
      let u;
      return e !== 0 && (u = q(e, n).slice(), a.__wbindgen_export4(e, n * 1, 1)), u;
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} fmt
   * @param {number} diagnostics_format
   * @returns {any}
   */
  get_artifact(t, e) {
    try {
      const o = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompileworld_get_artifact(o, this.__wbg_ptr, t, e);
      var n = l().getInt32(o + 0, !0), i = l().getInt32(o + 4, !0), s = l().getInt32(o + 8, !0);
      if (s)
        throw I(i);
      return I(n);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} kind
   * @param {string} selector
   * @param {string | null} [field]
   * @returns {string}
   */
  query(t, e, n) {
    let i, s;
    try {
      const L = a.__wbindgen_add_to_stack_pointer(-16), $ = T(e, a.__wbindgen_export, a.__wbindgen_export2), B = A;
      var o = O(n) ? 0 : T(n, a.__wbindgen_export, a.__wbindgen_export2), u = A;
      a.typstcompileworld_query(L, this.__wbg_ptr, t, $, B, o, u);
      var d = l().getInt32(L + 0, !0), g = l().getInt32(L + 4, !0), w = l().getInt32(L + 8, !0), m = l().getInt32(L + 12, !0), C = d, R = g;
      if (m)
        throw C = 0, R = 0, I(w);
      return i = C, s = R, q(C, R);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16), a.__wbindgen_export4(i, s, 1);
    }
  }
  /**
   * @param {IncrServer} state
   * @param {number} diagnostics_format
   * @returns {any}
   */
  incr_compile(t, e) {
    try {
      const o = a.__wbindgen_add_to_stack_pointer(-16);
      Lt(t, J), a.typstcompileworld_incr_compile(o, this.__wbg_ptr, t.__wbg_ptr, e);
      var n = l().getInt32(o + 0, !0), i = l().getInt32(o + 4, !0), s = l().getInt32(o + 8, !0);
      if (s)
        throw I(i);
      return I(n);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (ct.prototype[Symbol.dispose] = ct.prototype.free);
class ut {
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(ut.prototype);
    return e.__wbg_ptr = t, ae.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ae.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    a.__wbg_typstcompiler_free(t, 0);
  }
  reset() {
    try {
      const n = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompiler_reset(n, this.__wbg_ptr);
      var t = l().getInt32(n + 0, !0), e = l().getInt32(n + 4, !0);
      if (e)
        throw I(t);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {TypstFontResolver} fonts
   */
  set_fonts(t) {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      Lt(t, Q), a.typstcompiler_set_fonts(i, this.__wbg_ptr, t.__wbg_ptr);
      var e = l().getInt32(i + 0, !0), n = l().getInt32(i + 4, !0);
      if (n)
        throw I(e);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} inputs
   */
  set_inputs(t) {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompiler_set_inputs(i, this.__wbg_ptr, S(t));
      var e = l().getInt32(i + 0, !0), n = l().getInt32(i + 4, !0);
      if (n)
        throw I(e);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} path
   * @param {string} content
   * @returns {boolean}
   */
  add_source(t, e) {
    const n = T(t, a.__wbindgen_export, a.__wbindgen_export2), i = A, s = T(e, a.__wbindgen_export, a.__wbindgen_export2), o = A;
    return a.typstcompiler_add_source(this.__wbg_ptr, n, i, s, o) !== 0;
  }
  /**
   * @param {string} path
   * @param {Uint8Array} content
   * @returns {boolean}
   */
  map_shadow(t, e) {
    const n = T(t, a.__wbindgen_export, a.__wbindgen_export2), i = A, s = Ae(e, a.__wbindgen_export), o = A;
    return a.typstcompiler_map_shadow(this.__wbg_ptr, n, i, s, o) !== 0;
  }
  /**
   * @param {string} path
   * @returns {boolean}
   */
  unmap_shadow(t) {
    const e = T(t, a.__wbindgen_export, a.__wbindgen_export2), n = A;
    return a.typstcompiler_unmap_shadow(this.__wbg_ptr, e, n) !== 0;
  }
  reset_shadow() {
    a.typstcompiler_reset_shadow(this.__wbg_ptr);
  }
  /**
   * @returns {string[]}
   */
  get_loaded_fonts() {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompiler_get_loaded_fonts(i, this.__wbg_ptr);
      var t = l().getInt32(i + 0, !0), e = l().getInt32(i + 4, !0), n = Cr(t, e).slice();
      return a.__wbindgen_export4(t, e * 4, 4), n;
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} main_file_path
   * @returns {string}
   */
  get_ast(t) {
    let e, n;
    try {
      const w = a.__wbindgen_add_to_stack_pointer(-16), m = T(t, a.__wbindgen_export, a.__wbindgen_export2), C = A;
      a.typstcompiler_get_ast(w, this.__wbg_ptr, m, C);
      var i = l().getInt32(w + 0, !0), s = l().getInt32(w + 4, !0), o = l().getInt32(w + 8, !0), u = l().getInt32(w + 12, !0), d = i, g = s;
      if (u)
        throw d = 0, g = 0, I(o);
      return e = d, n = g, q(d, g);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16), a.__wbindgen_export4(e, n, 1);
    }
  }
  /**
   * @returns {any}
   */
  get_semantic_token_legend() {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompiler_get_semantic_token_legend(i, this.__wbg_ptr);
      var t = l().getInt32(i + 0, !0), e = l().getInt32(i + 4, !0), n = l().getInt32(i + 8, !0);
      if (n)
        throw I(e);
      return I(t);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} offset_encoding
   * @param {string | null} [file_path]
   * @param {string | null} [result_id]
   * @returns {object}
   */
  get_semantic_tokens(t, e, n) {
    try {
      const m = a.__wbindgen_add_to_stack_pointer(-16), C = T(t, a.__wbindgen_export, a.__wbindgen_export2), R = A;
      var i = O(e) ? 0 : T(e, a.__wbindgen_export, a.__wbindgen_export2), s = A, o = O(n) ? 0 : T(n, a.__wbindgen_export, a.__wbindgen_export2), u = A;
      a.typstcompiler_get_semantic_tokens(m, this.__wbg_ptr, C, R, i, s, o, u);
      var d = l().getInt32(m + 0, !0), g = l().getInt32(m + 4, !0), w = l().getInt32(m + 8, !0);
      if (w)
        throw I(g);
      return I(d);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [root]
   * @param {string | null} [main_file_path]
   * @param {(Array<any>)[] | null} [inputs]
   * @returns {TypstCompileWorld}
   */
  snapshot(t, e, n) {
    try {
      const R = a.__wbindgen_add_to_stack_pointer(-16);
      var i = O(t) ? 0 : T(t, a.__wbindgen_export, a.__wbindgen_export2), s = A, o = O(e) ? 0 : T(e, a.__wbindgen_export, a.__wbindgen_export2), u = A, d = O(n) ? 0 : ht(n, a.__wbindgen_export), g = A;
      a.typstcompiler_snapshot(R, this.__wbg_ptr, i, s, o, u, d, g);
      var w = l().getInt32(R + 0, !0), m = l().getInt32(R + 4, !0), C = l().getInt32(R + 8, !0);
      if (C)
        throw I(m);
      return ct.__wrap(w);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} fmt
   * @param {number} diagnostics_format
   * @returns {any}
   */
  get_artifact(t, e) {
    try {
      const o = a.__wbindgen_add_to_stack_pointer(-16), u = T(t, a.__wbindgen_export, a.__wbindgen_export2), d = A;
      a.typstcompiler_get_artifact(o, this.__wbg_ptr, u, d, e);
      var n = l().getInt32(o + 0, !0), i = l().getInt32(o + 4, !0), s = l().getInt32(o + 8, !0);
      if (s)
        throw I(i);
      return I(n);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null | undefined} main_file_path
   * @param {(Array<any>)[] | null | undefined} inputs
   * @param {string} fmt
   * @param {number} diagnostics_format
   * @returns {any}
   */
  compile(t, e, n, i) {
    try {
      const C = a.__wbindgen_add_to_stack_pointer(-16);
      var s = O(t) ? 0 : T(t, a.__wbindgen_export, a.__wbindgen_export2), o = A, u = O(e) ? 0 : ht(e, a.__wbindgen_export), d = A;
      const R = T(n, a.__wbindgen_export, a.__wbindgen_export2), L = A;
      a.typstcompiler_compile(C, this.__wbg_ptr, s, o, u, d, R, L, i);
      var g = l().getInt32(C + 0, !0), w = l().getInt32(C + 4, !0), m = l().getInt32(C + 8, !0);
      if (m)
        throw I(w);
      return I(g);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} main_file_path
   * @param {(Array<any>)[] | null | undefined} inputs
   * @param {string} selector
   * @param {string | null} [field]
   * @returns {string}
   */
  query(t, e, n, i) {
    let s, o;
    try {
      const z = a.__wbindgen_add_to_stack_pointer(-16), Le = T(t, a.__wbindgen_export, a.__wbindgen_export2), $e = A;
      var u = O(e) ? 0 : ht(e, a.__wbindgen_export), d = A;
      const De = T(n, a.__wbindgen_export, a.__wbindgen_export2), We = A;
      var g = O(i) ? 0 : T(i, a.__wbindgen_export, a.__wbindgen_export2), w = A;
      a.typstcompiler_query(z, this.__wbg_ptr, Le, $e, u, d, De, We, g, w);
      var m = l().getInt32(z + 0, !0), C = l().getInt32(z + 4, !0), R = l().getInt32(z + 8, !0), L = l().getInt32(z + 12, !0), $ = m, B = C;
      if (L)
        throw $ = 0, B = 0, I(R);
      return s = $, o = B, q($, B);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16), a.__wbindgen_export4(s, o, 1);
    }
  }
  /**
   * @returns {IncrServer}
   */
  create_incr_server() {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompiler_create_incr_server(i, this.__wbg_ptr);
      var t = l().getInt32(i + 0, !0), e = l().getInt32(i + 4, !0), n = l().getInt32(i + 8, !0);
      if (n)
        throw I(e);
      return J.__wrap(t);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} main_file_path
   * @param {(Array<any>)[] | null | undefined} inputs
   * @param {IncrServer} state
   * @param {number} diagnostics_format
   * @returns {any}
   */
  incr_compile(t, e, n, i) {
    try {
      const w = a.__wbindgen_add_to_stack_pointer(-16), m = T(t, a.__wbindgen_export, a.__wbindgen_export2), C = A;
      var s = O(e) ? 0 : ht(e, a.__wbindgen_export), o = A;
      Lt(n, J), a.typstcompiler_incr_compile(w, this.__wbg_ptr, m, C, s, o, n.__wbg_ptr, i);
      var u = l().getInt32(w + 0, !0), d = l().getInt32(w + 4, !0), g = l().getInt32(w + 8, !0);
      if (g)
        throw I(d);
      return I(u);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (ut.prototype[Symbol.dispose] = ut.prototype.free);
class Wt {
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ce.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    a.__wbg_typstcompilerbuilder_free(t, 0);
  }
  constructor() {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompilerbuilder_new(i);
      var t = l().getInt32(i + 0, !0), e = l().getInt32(i + 4, !0), n = l().getInt32(i + 8, !0);
      if (n)
        throw I(e);
      return this.__wbg_ptr = t >>> 0, ce.register(this, this.__wbg_ptr, this), this;
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  set_dummy_access_model() {
    try {
      const n = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstcompilerbuilder_set_dummy_access_model(n, this.__wbg_ptr);
      var t = l().getInt32(n + 0, !0), e = l().getInt32(n + 4, !0);
      if (e)
        throw I(t);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} context
   * @param {Function} mtime_fn
   * @param {Function} is_file_fn
   * @param {Function} real_path_fn
   * @param {Function} read_all_fn
   * @returns {Promise<void>}
   */
  set_access_model(t, e, n, i, s) {
    const o = a.typstcompilerbuilder_set_access_model(this.__wbg_ptr, S(t), S(e), S(n), S(i), S(s));
    return I(o);
  }
  /**
   * @param {any} context
   * @param {Function} real_resolve_fn
   * @returns {Promise<void>}
   */
  set_package_registry(t, e) {
    const n = a.typstcompilerbuilder_set_package_registry(this.__wbg_ptr, S(t), S(e));
    return I(n);
  }
  /**
   * @param {Uint8Array} data
   * @returns {Promise<void>}
   */
  add_raw_font(t) {
    const e = a.typstcompilerbuilder_add_raw_font(this.__wbg_ptr, S(t));
    return I(e);
  }
  /**
   * @param {any} font
   * @param {Function} blob
   * @returns {Promise<void>}
   */
  add_lazy_font(t, e) {
    const n = a.typstcompilerbuilder_add_lazy_font(this.__wbg_ptr, S(t), S(e));
    return I(n);
  }
  /**
   * @returns {Promise<TypstCompiler>}
   */
  build() {
    const t = this.__destroy_into_raw(), e = a.typstcompilerbuilder_build(t);
    return I(e);
  }
}
Symbol.dispose && (Wt.prototype[Symbol.dispose] = Wt.prototype.free);
class Q {
  static __wrap(t) {
    t = t >>> 0;
    const e = Object.create(Q.prototype);
    return e.__wbg_ptr = t, ue.register(e, e.__wbg_ptr, e), e;
  }
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ue.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    a.__wbg_typstfontresolver_free(t, 0);
  }
}
Symbol.dispose && (Q.prototype[Symbol.dispose] = Q.prototype.free);
class jt {
  __destroy_into_raw() {
    const t = this.__wbg_ptr;
    return this.__wbg_ptr = 0, de.unregister(this), t;
  }
  free() {
    const t = this.__destroy_into_raw();
    a.__wbg_typstfontresolverbuilder_free(t, 0);
  }
  constructor() {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstfontresolverbuilder_new(i);
      var t = l().getInt32(i + 0, !0), e = l().getInt32(i + 4, !0), n = l().getInt32(i + 8, !0);
      if (n)
        throw I(e);
      return this.__wbg_ptr = t >>> 0, de.register(this, this.__wbg_ptr, this), this;
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {Uint8Array} buffer
   * @returns {any}
   */
  get_font_info(t) {
    try {
      const s = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstfontresolverbuilder_get_font_info(s, this.__wbg_ptr, S(t));
      var e = l().getInt32(s + 0, !0), n = l().getInt32(s + 4, !0), i = l().getInt32(s + 8, !0);
      if (i)
        throw I(n);
      return I(e);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * Adds font data to the searcher.
   * @param {Uint8Array} buffer
   */
  add_raw_font(t) {
    try {
      const i = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstfontresolverbuilder_add_raw_font(i, this.__wbg_ptr, S(t));
      var e = l().getInt32(i + 0, !0), n = l().getInt32(i + 4, !0);
      if (n)
        throw I(e);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * Adds callback that loads font data lazily to the searcher.
   * `get_font_info` can be used to get the font info.
   * @param {any} font
   * @param {Function} blob
   */
  add_lazy_font(t, e) {
    try {
      const s = a.__wbindgen_add_to_stack_pointer(-16);
      a.typstfontresolverbuilder_add_lazy_font(s, this.__wbg_ptr, S(t), S(e));
      var n = l().getInt32(s + 0, !0), i = l().getInt32(s + 4, !0);
      if (i)
        throw I(n);
    } finally {
      a.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @returns {Promise<TypstFontResolver>}
   */
  build() {
    const t = this.__destroy_into_raw(), e = a.typstfontresolverbuilder_build(t);
    return I(e);
  }
}
Symbol.dispose && (jt.prototype[Symbol.dispose] = jt.prototype.free);
function Dr(r) {
  const t = a.get_font_info(S(r));
  return I(t);
}
const Wr = /* @__PURE__ */ new Set(["basic", "cors", "default"]);
async function jr(r, t) {
  if (typeof Response == "function" && r instanceof Response) {
    if (typeof WebAssembly.instantiateStreaming == "function")
      try {
        return await WebAssembly.instantiateStreaming(r, t);
      } catch (n) {
        if (r.ok && Wr.has(r.type) && r.headers.get("Content-Type") !== "application/wasm")
          console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve Wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", n);
        else
          throw n;
      }
    const e = await r.arrayBuffer();
    return await WebAssembly.instantiate(e, t);
  } else {
    const e = await WebAssembly.instantiate(r, t);
    return e instanceof WebAssembly.Instance ? { instance: e, module: r } : e;
  }
}
function Ce() {
  const r = {};
  return r.wbg = {}, r.wbg.__wbg_Error_52673b7de5a0ca89 = function(t, e) {
    const n = Error(q(t, e));
    return S(n);
  }, r.wbg.__wbg_Number_2d1dcfcf4ec51736 = function(t) {
    return Number(y(t));
  }, r.wbg.__wbg___wbindgen_bigint_get_as_i64_6e32f5e6aff02e1d = function(t, e) {
    const n = y(e), i = typeof n == "bigint" ? n : void 0;
    l().setBigInt64(t + 8, O(i) ? BigInt(0) : i, !0), l().setInt32(t + 0, !O(i), !0);
  }, r.wbg.__wbg___wbindgen_boolean_get_dea25b33882b895b = function(t) {
    const e = y(t), n = typeof e == "boolean" ? e : void 0;
    return O(n) ? 16777215 : n ? 1 : 0;
  }, r.wbg.__wbg___wbindgen_debug_string_adfb662ae34724b6 = function(t, e) {
    const n = $t(y(e)), i = T(n, a.__wbindgen_export, a.__wbindgen_export2), s = A;
    l().setInt32(t + 4, s, !0), l().setInt32(t + 0, i, !0);
  }, r.wbg.__wbg___wbindgen_in_0d3e1e8f0c669317 = function(t, e) {
    return y(t) in y(e);
  }, r.wbg.__wbg___wbindgen_is_bigint_0e1a2e3f55cfae27 = function(t) {
    return typeof y(t) == "bigint";
  }, r.wbg.__wbg___wbindgen_is_function_8d400b8b1af978cd = function(t) {
    return typeof y(t) == "function";
  }, r.wbg.__wbg___wbindgen_is_object_ce774f3490692386 = function(t) {
    const e = y(t);
    return typeof e == "object" && e !== null;
  }, r.wbg.__wbg___wbindgen_is_string_704ef9c8fc131030 = function(t) {
    return typeof y(t) == "string";
  }, r.wbg.__wbg___wbindgen_is_undefined_f6b95eab589e0269 = function(t) {
    return y(t) === void 0;
  }, r.wbg.__wbg___wbindgen_jsval_eq_b6101cc9cef1fe36 = function(t, e) {
    return y(t) === y(e);
  }, r.wbg.__wbg___wbindgen_jsval_loose_eq_766057600fdd1b0d = function(t, e) {
    return y(t) == y(e);
  }, r.wbg.__wbg___wbindgen_number_get_9619185a74197f95 = function(t, e) {
    const n = y(e), i = typeof n == "number" ? n : void 0;
    l().setFloat64(t + 8, O(i) ? 0 : i, !0), l().setInt32(t + 0, !O(i), !0);
  }, r.wbg.__wbg___wbindgen_string_get_a2a31e16edf96e42 = function(t, e) {
    const n = y(e), i = typeof n == "string" ? n : void 0;
    var s = O(i) ? 0 : T(i, a.__wbindgen_export, a.__wbindgen_export2), o = A;
    l().setInt32(t + 4, o, !0), l().setInt32(t + 0, s, !0);
  }, r.wbg.__wbg___wbindgen_throw_dd24417ed36fc46e = function(t, e) {
    throw new Error(q(t, e));
  }, r.wbg.__wbg__wbg_cb_unref_87dfb5aaa0cbcea7 = function(t) {
    y(t)._wbg_cb_unref();
  }, r.wbg.__wbg_call_3020136f7a2d6e44 = function() {
    return Y(function(t, e, n) {
      const i = y(t).call(y(e), y(n));
      return S(i);
    }, arguments);
  }, r.wbg.__wbg_call_78f94eb02ec7f9b2 = function() {
    return Y(function(t, e, n, i, s) {
      const o = y(t).call(y(e), y(n), y(i), y(s));
      return S(o);
    }, arguments);
  }, r.wbg.__wbg_call_abb4ff46ce38be40 = function() {
    return Y(function(t, e) {
      const n = y(t).call(y(e));
      return S(n);
    }, arguments);
  }, r.wbg.__wbg_done_62ea16af4ce34b24 = function(t) {
    return y(t).done;
  }, r.wbg.__wbg_entries_83c79938054e065f = function(t) {
    const e = Object.entries(y(t));
    return S(e);
  }, r.wbg.__wbg_error_7534b8e9a36f1ab4 = function(t, e) {
    let n, i;
    try {
      n = t, i = e, console.error(q(t, e));
    } finally {
      a.__wbindgen_export4(n, i, 1);
    }
  }, r.wbg.__wbg_error_85faeb8919b11cc6 = function(t, e, n) {
    console.error(y(t), y(e), y(n));
  }, r.wbg.__wbg_getTimezoneOffset_45389e26d6f46823 = function(t) {
    return y(t).getTimezoneOffset();
  }, r.wbg.__wbg_get_6b7bd52aca3f9671 = function(t, e) {
    const n = y(t)[e >>> 0];
    return S(n);
  }, r.wbg.__wbg_get_af9dab7e9603ea93 = function() {
    return Y(function(t, e) {
      const n = Reflect.get(y(t), y(e));
      return S(n);
    }, arguments);
  }, r.wbg.__wbg_get_with_ref_key_1dc361bd10053bfe = function(t, e) {
    const n = y(t)[y(e)];
    return S(n);
  }, r.wbg.__wbg_info_ce6bcc489c22f6f0 = function(t) {
    console.info(y(t));
  }, r.wbg.__wbg_instanceof_ArrayBuffer_f3320d2419cd0355 = function(t) {
    let e;
    try {
      e = y(t) instanceof ArrayBuffer;
    } catch {
      e = !1;
    }
    return e;
  }, r.wbg.__wbg_instanceof_Map_084be8da74364158 = function(t) {
    let e;
    try {
      e = y(t) instanceof Map;
    } catch {
      e = !1;
    }
    return e;
  }, r.wbg.__wbg_instanceof_Uint8Array_da54ccc9d3e09434 = function(t) {
    let e;
    try {
      e = y(t) instanceof Uint8Array;
    } catch {
      e = !1;
    }
    return e;
  }, r.wbg.__wbg_isArray_51fd9e6422c0a395 = function(t) {
    return Array.isArray(y(t));
  }, r.wbg.__wbg_isSafeInteger_ae7d3f054d55fa16 = function(t) {
    return Number.isSafeInteger(y(t));
  }, r.wbg.__wbg_iterator_27b7c8b35ab3e86b = function() {
    return S(Symbol.iterator);
  }, r.wbg.__wbg_length_22ac23eaec9d8053 = function(t) {
    return y(t).length;
  }, r.wbg.__wbg_length_d45040a40c570362 = function(t) {
    return y(t).length;
  }, r.wbg.__wbg_new_1ba21ce319a06297 = function() {
    const t = new Object();
    return S(t);
  }, r.wbg.__wbg_new_25f239778d6112b9 = function() {
    const t = new Array();
    return S(t);
  }, r.wbg.__wbg_new_6421f6084cc5bc5a = function(t) {
    const e = new Uint8Array(y(t));
    return S(e);
  }, r.wbg.__wbg_new_8a6f238a6ece86ea = function() {
    const t = new Error();
    return S(t);
  }, r.wbg.__wbg_new_b2db8aa2650f793a = function(t) {
    const e = new Date(y(t));
    return S(e);
  }, r.wbg.__wbg_new_df1173567d5ff028 = function(t, e) {
    const n = new Error(q(t, e));
    return S(n);
  }, r.wbg.__wbg_new_ff12d2b041fb48f1 = function(t, e) {
    try {
      var n = { a: t, b: e }, i = (o, u) => {
        const d = n.a;
        n.a = 0;
        try {
          return $r(d, n.b, o, u);
        } finally {
          n.a = d;
        }
      };
      const s = new Promise(i);
      return S(s);
    } finally {
      n.a = n.b = 0;
    }
  }, r.wbg.__wbg_new_from_slice_db0691b69e9d3891 = function(t, e) {
    const n = new Uint32Array(Fr(t, e));
    return S(n);
  }, r.wbg.__wbg_new_from_slice_f9c22b9153b26992 = function(t, e) {
    const n = new Uint8Array(Dt(t, e));
    return S(n);
  }, r.wbg.__wbg_new_no_args_cb138f77cf6151ee = function(t, e) {
    const n = new Function(q(t, e));
    return S(n);
  }, r.wbg.__wbg_new_with_args_df9e7125ffe55248 = function(t, e, n, i) {
    const s = new Function(q(t, e), q(n, i));
    return S(s);
  }, r.wbg.__wbg_next_138a17bbf04e926c = function(t) {
    const e = y(t).next;
    return S(e);
  }, r.wbg.__wbg_next_3cfe5c0fe2a4cc53 = function() {
    return Y(function(t) {
      const e = y(t).next();
      return S(e);
    }, arguments);
  }, r.wbg.__wbg_now_69d776cd24f5215b = function() {
    return Date.now();
  }, r.wbg.__wbg_prototypesetcall_dfe9b766cdc1f1fd = function(t, e, n) {
    Uint8Array.prototype.set.call(Dt(t, e), y(n));
  }, r.wbg.__wbg_proxycontext_new = function(t) {
    const e = at.__wrap(t);
    return S(e);
  }, r.wbg.__wbg_push_7d9be8f38fc13975 = function(t, e) {
    return y(t).push(y(e));
  }, r.wbg.__wbg_queueMicrotask_9b549dfce8865860 = function(t) {
    const e = y(t).queueMicrotask;
    return S(e);
  }, r.wbg.__wbg_queueMicrotask_fca69f5bfad613a5 = function(t) {
    queueMicrotask(y(t));
  }, r.wbg.__wbg_resolve_fd5bfbaa4ce36e1e = function(t) {
    const e = Promise.resolve(y(t));
    return S(e);
  }, r.wbg.__wbg_set_3f1d0b984ed272ed = function(t, e, n) {
    y(t)[I(e)] = I(n);
  }, r.wbg.__wbg_set_781438a03c0c3c81 = function() {
    return Y(function(t, e, n) {
      return Reflect.set(y(t), y(e), y(n));
    }, arguments);
  }, r.wbg.__wbg_set_7df433eea03a5c14 = function(t, e, n) {
    y(t)[e >>> 0] = I(n);
  }, r.wbg.__wbg_stack_0ed75d68575b0f3c = function(t, e) {
    const n = y(e).stack, i = T(n, a.__wbindgen_export, a.__wbindgen_export2), s = A;
    l().setInt32(t + 4, s, !0), l().setInt32(t + 0, i, !0);
  }, r.wbg.__wbg_static_accessor_GLOBAL_769e6b65d6557335 = function() {
    const t = typeof global > "u" ? null : global;
    return O(t) ? 0 : S(t);
  }, r.wbg.__wbg_static_accessor_GLOBAL_THIS_60cf02db4de8e1c1 = function() {
    const t = typeof globalThis > "u" ? null : globalThis;
    return O(t) ? 0 : S(t);
  }, r.wbg.__wbg_static_accessor_SELF_08f5a74c69739274 = function() {
    const t = typeof self > "u" ? null : self;
    return O(t) ? 0 : S(t);
  }, r.wbg.__wbg_static_accessor_WINDOW_a8924b26aa92d024 = function() {
    const t = typeof window > "u" ? null : window;
    return O(t) ? 0 : S(t);
  }, r.wbg.__wbg_then_4f95312d68691235 = function(t, e) {
    const n = y(t).then(y(e));
    return S(n);
  }, r.wbg.__wbg_typstcompiler_new = function(t) {
    const e = ut.__wrap(t);
    return S(e);
  }, r.wbg.__wbg_typstfontresolver_new = function(t) {
    const e = Q.__wrap(t);
    return S(e);
  }, r.wbg.__wbg_value_57b7b035e117f7ee = function(t) {
    const e = y(t).value;
    return S(e);
  }, r.wbg.__wbindgen_cast_2241b6af4c4b2941 = function(t, e) {
    const n = q(t, e);
    return S(n);
  }, r.wbg.__wbindgen_cast_3334ea73b4b28ba3 = function(t, e) {
    const n = Pr(t, e, a.__wasm_bindgen_func_elem_957, Lr);
    return S(n);
  }, r.wbg.__wbindgen_cast_4625c577ab2ec9ee = function(t) {
    const e = BigInt.asUintN(64, t);
    return S(e);
  }, r.wbg.__wbindgen_cast_9ae0607507abb057 = function(t) {
    return S(t);
  }, r.wbg.__wbindgen_cast_d6cd19b81560fd6e = function(t) {
    return S(t);
  }, r.wbg.__wbindgen_object_clone_ref = function(t) {
    const e = y(t);
    return S(e);
  }, r.wbg.__wbindgen_object_drop_ref = function(t) {
    I(t);
  }, r;
}
function Fe(r, t) {
  return a = r.exports, Me.__wbindgen_wasm_module = t, G = null, ft = null, wt = null, a;
}
function Br(r) {
  if (a !== void 0) return a;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module: r } = r : console.warn("using deprecated parameters for `initSync()`; pass a single object instead"));
  const t = Ce();
  r instanceof WebAssembly.Module || (r = new WebAssembly.Module(r));
  const e = new WebAssembly.Instance(r, t);
  return Fe(e, r);
}
async function Me(r) {
  if (a !== void 0) return a;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module_or_path: r } = r : console.warn("using deprecated parameters for the initialization function; pass a single object instead")), typeof r > "u" && (r = Pe("typst_ts_web_compiler_bg.wasm", import.meta.url));
  const t = Ce();
  (typeof r == "string" || typeof Request == "function" && r instanceof Request || typeof URL == "function" && r instanceof URL) && (r = fetch(r));
  const { instance: e, module: n } = await jr(await r, t);
  return Fe(e, n);
}
let Pe = async function(r, t) {
  throw new Error("Cannot import wasm module without importer: " + r + " " + t);
};
function Te(r) {
  Pe = r;
}
let zr = async function(r, t) {
  const e = new Function("m", "return import(m)"), { readFileSync: n } = await e("fs"), i = new URL(r, t);
  return await n(i).buffer;
};
const qr = typeof process < "u" && process.versions != null && process.versions.node != null;
qr && Te(zr);
const Oe = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  IncrServer: J,
  ProxyContext: at,
  TypstCompileWorld: ct,
  TypstCompiler: ut,
  TypstCompilerBuilder: Wt,
  TypstFontResolver: Q,
  TypstFontResolverBuilder: jt,
  default: Me,
  get_font_info: Dr,
  initSync: Br,
  setImportWasmModule: Te
}, Symbol.toStringTag, { value: "Module" }));
export {
  ar as FetchAccessModel,
  ur as FetchPackageRegistry,
  ve as createTypstCompiler,
  Ut as createTypstRenderer,
  Z as loadFonts,
  Nr as preloadRemoteFonts,
  qt as preloadSystemFonts,
  Xr as rendererBuildInfo
};
